
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Penetrance &#8212; simuPOP  documentation</title>
    <link rel="stylesheet" href="_static/simuPOP.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Quantitative Trait" href="refManual_ch3_sec8.html" />
    <link rel="prev" title="Mutation" href="refManual_ch3_sec6.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="penetrance">
<h1>Penetrance<a class="headerlink" href="#penetrance" title="Permalink to this headline">¶</a></h1>
<div class="section" id="class-basepenetrance">
<span id="index-0"></span><h2>class BasePenetrance<a class="headerlink" href="#class-basepenetrance" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="BasePenetrance">
<em class="property">class </em><code class="descname">BasePenetrance</code><a class="headerlink" href="#BasePenetrance" title="Permalink to this definition">¶</a></dt>
<dd><p>A penetrance model models the probability that an individual has a
certain disease provided that he or she has certain genetic
(genotype) and environmental (information field) riske factors. A
penetrance operator calculates this probability according to
provided information and set his or her affection status randomly.
For example, an individual will have probability 0.8 to be affected
if the penetrance is 0.8. This class is the base class to all
penetrance operators and defines a common interface for all
penetrance operators.</p>
<p>A penetrance operator can be applied at any stage of an
evolutionary cycle. If it is applied before or after mating, it
will set affection status of all parents and offspring,
respectively. If it is applied during mating, it will set the
affection status of each offspring. You can also apply a penetrance
operator to an individual using its <code class="docutils literal notranslate"><span class="pre">applyToIndividual</span></code> member
function.</p>
<p>By default, a penetrance operator assigns affection status of
individuals but does not save the actual penetrance value. However,
if an information field is specified, penetrance values will be
saved to this field for future analysis.</p>
<p>When a penetrance operator is applied to a population, it is only
applied to the current generation. You can, however, use parameter
<em>ancGens</em> to set affection status for all ancestral generations
(<code class="docutils literal notranslate"><span class="pre">ALL_AVAIL</span></code>), or individuals in specified generations if a list
of ancestral generations is specified. Note that this parameter is
ignored if the operator is applied during mating.</p>
<dl class="method">
<dt id="BasePenetrance.BasePenetrance">
<code class="descname">BasePenetrance</code><span class="sig-paren">(</span><em>ancGens=UNSPECIFIED</em>, <em>begin=0</em>, <em>end=-1</em>, <em>step=1</em>, <em>at=[]</em>, <em>reps=ALL_AVAIL</em>, <em>subPops=ALL_AVAIL</em>, <em>infoFields=[]</em><span class="sig-paren">)</span><a class="headerlink" href="#BasePenetrance.BasePenetrance" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a base penetrance operator. This operator assign
individual affection status in the present generation (default).
If <code class="docutils literal notranslate"><span class="pre">ALL_AVAIL</span></code> or a list of ancestral generations are spcified
in parameter <em>ancGens</em>, individuals in specified ancestral
generations will be processed. A penetrance operator can be
applied to specified (virtual) subpopulations (parameter
<em>subPops</em>) and replicates (parameter <em>reps</em>). If an informatio
field is given, penetrance value will be stored in this
information field of each individual.</p>
</dd></dl>

<dl class="method">
<dt id="BasePenetrance.apply">
<code class="descname">apply</code><span class="sig-paren">(</span><em>pop</em><span class="sig-paren">)</span><a class="headerlink" href="#BasePenetrance.apply" title="Permalink to this definition">¶</a></dt>
<dd><p>set penetrance to all individuals and record penetrance if
requested</p>
</dd></dl>

<dl class="method">
<dt id="BasePenetrance.applyToIndividual">
<code class="descname">applyToIndividual</code><span class="sig-paren">(</span><em>ind</em>, <em>pop=None</em><span class="sig-paren">)</span><a class="headerlink" href="#BasePenetrance.applyToIndividual" title="Permalink to this definition">¶</a></dt>
<dd><p>Apply the penetrance operator to a single individual <em>ind</em> and
set his or her affection status. A population reference can be
passed if the penetrance model depends on population properties
such as generation number. This function returns the affection
status.</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="class-mappenetrance">
<h2>class MapPenetrance<a class="headerlink" href="#class-mappenetrance" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="MapPenetrance">
<em class="property">class </em><code class="descname">MapPenetrance</code><a class="headerlink" href="#MapPenetrance" title="Permalink to this definition">¶</a></dt>
<dd><p>This penetrance operator assigns individual affection status using
a user-specified penetrance dictionary.</p>
<dl class="method">
<dt id="MapPenetrance.MapPenetrance">
<code class="descname">MapPenetrance</code><span class="sig-paren">(</span><em>loci</em>, <em>penetrance</em>, <em>ancGens=UNSPECIFIED</em>, <em>begin=0</em>, <em>end=-1</em>, <em>step=1</em>, <em>at=[]</em>, <em>reps=ALL_AVAIL</em>, <em>subPops=ALL_AVAIL</em>, <em>infoFields=[]</em><span class="sig-paren">)</span><a class="headerlink" href="#MapPenetrance.MapPenetrance" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a penetrance operator that get penetrance value from a
dictionary <em>penetrance</em> with genotype at <em>loci</em> as keys, and
<em>penetrance</em> as values. For each individual, genotypes at <em>loci</em>
are collected one by one (e.g. p0_loc0, p1_loc0, p0_loc1,
p1_loc1… for a diploid individual) and are looked up in the
dictionary. Parameter <em>loci</em> can be a list of loci indexes,
names, list of chromosome position pairs, <code class="docutils literal notranslate"><span class="pre">ALL_AVAIL</span></code>, or a
function with optional parameter <code class="docutils literal notranslate"><span class="pre">pop</span></code> that will be called at
each ganeeration to determine indexes of loci. If a genotype
cannot be found, it will be looked up again without phase
information (e.g. <code class="docutils literal notranslate"><span class="pre">(1,0)</span></code> will match key <code class="docutils literal notranslate"><span class="pre">(0,1)</span></code>). If the
genotype still can not be found, a <code class="xref py py-class docutils literal notranslate"><span class="pre">ValueError</span></code> will be
raised. This operator supports sex chromosomes and haplodiploid
populations. In these cases, only valid genotypes should be used
to generator the dictionary keys.</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="class-mapenetrance">
<h2>class MaPenetrance<a class="headerlink" href="#class-mapenetrance" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="MaPenetrance">
<em class="property">class </em><code class="descname">MaPenetrance</code><a class="headerlink" href="#MaPenetrance" title="Permalink to this definition">¶</a></dt>
<dd><p>This operator is called a ‘multi-allele’ penetrance operator
because it groups multiple alleles into two groups: wildtype and
non-wildtype alleles. Alleles in each allele group are assumed to
have the same effect on individual penetrance. If we denote all
wildtype alleles as <code class="docutils literal notranslate"><span class="pre">A</span></code>, and all non-wildtype alleles <code class="docutils literal notranslate"><span class="pre">a</span></code>, this
operator assign  Individual penetrance according to genotype
<code class="docutils literal notranslate"><span class="pre">AA</span></code>, <code class="docutils literal notranslate"><span class="pre">Aa</span></code>, <code class="docutils literal notranslate"><span class="pre">aa</span></code> in the diploid case, and <code class="docutils literal notranslate"><span class="pre">A</span></code> and <code class="docutils literal notranslate"><span class="pre">a</span></code> in
the haploid case.</p>
<dl class="method">
<dt id="MaPenetrance.MaPenetrance">
<code class="descname">MaPenetrance</code><span class="sig-paren">(</span><em>loci</em>, <em>penetrance</em>, <em>wildtype=0</em>, <em>ancGens=UNSPECIFIED</em>, <em>begin=0</em>, <em>end=-1</em>, <em>step=1</em>, <em>at=[]</em>, <em>reps=ALL_AVAIL</em>, <em>subPops=ALL_AVAIL</em>, <em>infoFields=[]</em><span class="sig-paren">)</span><a class="headerlink" href="#MaPenetrance.MaPenetrance" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates a multi-allele penetrance operator that groups multiple
alleles into a wildtype group (with alleles <em>wildtype</em>, default
to <code class="docutils literal notranslate"><span class="pre">[0]</span></code>), and a non-wildtype group. A list of penetrance
values is specified through parameter <em>penetrance</em>, for
genotypes at one or more <em>loci</em>. Parameter <em>loci</em> can be a list
of loci indexes, names, list of chromosome position pairs,
<code class="docutils literal notranslate"><span class="pre">ALL_AVAIL</span></code>, or a function with optional parameter <code class="docutils literal notranslate"><span class="pre">pop</span></code>
that will be called at each ganeeration to determine indexes of
loci. If we denote wildtype alleles using capital letters <code class="docutils literal notranslate"><span class="pre">A</span></code>,
<code class="docutils literal notranslate"><span class="pre">B</span></code> … and non-wildtype alleles using small letters <code class="docutils literal notranslate"><span class="pre">a</span></code>,
<code class="docutils literal notranslate"><span class="pre">b</span></code> …, the penetrance values should be for</p>
<ul class="simple">
<li>genotypes <code class="docutils literal notranslate"><span class="pre">A</span></code> and <code class="docutils literal notranslate"><span class="pre">a</span></code> for the haploid single-locus case,</li>
<li>genotypes <code class="docutils literal notranslate"><span class="pre">AB</span></code>, <code class="docutils literal notranslate"><span class="pre">Ab</span></code>, <code class="docutils literal notranslate"><span class="pre">aB</span></code> and <code class="docutils literal notranslate"><span class="pre">bb</span></code> for haploid
two=locus cases,</li>
<li>genotypes <code class="docutils literal notranslate"><span class="pre">AA</span></code>, <code class="docutils literal notranslate"><span class="pre">Aa</span></code> and <code class="docutils literal notranslate"><span class="pre">aa</span></code> for diploid single-locus
cases,</li>
<li>genotypes <code class="docutils literal notranslate"><span class="pre">AABB</span></code>, <code class="docutils literal notranslate"><span class="pre">AABb</span></code>, <code class="docutils literal notranslate"><span class="pre">AAbb</span></code>, <code class="docutils literal notranslate"><span class="pre">AaBB</span></code>, <code class="docutils literal notranslate"><span class="pre">AaBb</span></code>,
<code class="docutils literal notranslate"><span class="pre">Aabb</span></code>, <code class="docutils literal notranslate"><span class="pre">aaBB</span></code>, <code class="docutils literal notranslate"><span class="pre">aaBb</span></code>, and <code class="docutils literal notranslate"><span class="pre">aabb</span></code> for diploid two-
locus cases,</li>
<li>and in general 2**n for diploid and 3**n for haploid cases if
there are <code class="docutils literal notranslate"><span class="pre">n</span></code> loci.</li>
</ul>
<p>This operator does not support haplodiploid populations and sex
chromosomes.</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="class-mlpenetrance">
<h2>class MlPenetrance<a class="headerlink" href="#class-mlpenetrance" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="MlPenetrance">
<em class="property">class </em><code class="descname">MlPenetrance</code><a class="headerlink" href="#MlPenetrance" title="Permalink to this definition">¶</a></dt>
<dd><p>This penetrance operator is created by a list of penetrance
operators. When it is applied to an individual, it applies these
penetrance operators to the individual, obtain a list of penetrance
values, and compute a combined penetrance value from them and
assign affection status accordingly. ADDITIVE, multiplicative, and
a heterogeneour multi-locus model are supported. Please refer to
Neil Rish (1989) “Linkage Strategies for</p>
<p>Genetically Complex Traits” for some analysis of these models.</p>
<dl class="method">
<dt id="MlPenetrance.MlPenetrance">
<code class="descname">MlPenetrance</code><span class="sig-paren">(</span><em>ops</em>, <em>mode=MULTIPLICATIVE</em>, <em>ancGens=UNSPECIFIED</em>, <em>begin=0</em>, <em>end=-1</em>, <em>step=1</em>, <em>at=[]</em>, <em>reps=ALL_AVAIL</em>, <em>subPops=ALL_AVAIL</em>, <em>infoFields=[]</em><span class="sig-paren">)</span><a class="headerlink" href="#MlPenetrance.MlPenetrance" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a multiple-locus penetrance operator from a list
penetrance operator <em>ops</em>. When this operator is applied to an
individual (parents when used before mating and offspring when
used during mating), it applies these operators to the
individual and obtain a list of (usually single-locus)
penetrance values. These penetrance values are combined to a
single penetrance value using</p>
<ul class="simple">
<li><em>Prod(f_i)</em>, namely the product of individual penetrance if
<em>mode</em> = <code class="docutils literal notranslate"><span class="pre">MULTIPLICATIVE</span></code>,</li>
<li><em>sum(f_i)</em> if <em>mode</em> = <code class="docutils literal notranslate"><span class="pre">ADDITIVE</span></code>, and</li>
<li><em>1-Prod(1 - f_i)</em> if <em>mode</em> = <code class="docutils literal notranslate"><span class="pre">HETEROGENEITY</span></code></li>
</ul>
<p>0 or 1 will be returned if the combined penetrance value is less
than zero or greater than 1.</p>
<p>Applicability parameters (begin, end, step, at, reps, subPops)
could be used in both <a class="reference internal" href="refManual_ch3_sec9.html#MlSelector" title="MlSelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">MlSelector</span></code></a> and selectors in
parameter <em>ops</em>, but parameters in <a class="reference internal" href="refManual_ch3_sec9.html#MlSelector" title="MlSelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">MlSelector</span></code></a> will be
interpreted first.</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="class-pypenetrance">
<h2>class PyPenetrance<a class="headerlink" href="#class-pypenetrance" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="PyPenetrance">
<em class="property">class </em><code class="descname">PyPenetrance</code><a class="headerlink" href="#PyPenetrance" title="Permalink to this definition">¶</a></dt>
<dd><p>This penetrance operator assigns penetrance values by calling a
user provided function. It accepts a list of loci (parameter
<code class="docutils literal notranslate"><span class="pre">loci</span></code>), and a Python function <code class="docutils literal notranslate"><span class="pre">func</span></code> which should be defined
with one or more of parameters <code class="docutils literal notranslate"><span class="pre">geno</span></code>, <code class="docutils literal notranslate"><span class="pre">mut</span></code>, <code class="docutils literal notranslate"><span class="pre">gen</span></code>, <code class="docutils literal notranslate"><span class="pre">ind</span></code>,
<code class="docutils literal notranslate"><span class="pre">pop</span></code>, or names of information fields. When this operator is
applied to a population, it passes genotypes or mutants (non-zero
alleles) at specified loci at specified loci, generation number, a
reference to an individual, a reference to the current population
(usually used to retrieve population variables) and values at
specified information fields to respective parameters of this
function. Genotypes of each individual are passed as a tuple of
alleles arranged locus by locus (in the order of A1,A2,B1,B2 for
loci A and B). Mutants are passed as a default dictionary of loci
index (with respect to all genotype of individuals, not just the
first ploidy) and alleles. The returned penetrance values will be
used to determine the affection status of each individual.</p>
<dl class="method">
<dt id="PyPenetrance.PyPenetrance">
<code class="descname">PyPenetrance</code><span class="sig-paren">(</span><em>func</em>, <em>loci=[]</em>, <em>ancGens=UNSPECIFIED</em>, <em>begin=0</em>, <em>end=-1</em>, <em>step=1</em>, <em>at=[]</em>, <em>reps=ALL_AVAIL</em>, <em>subPops=ALL_AVAIL</em>, <em>infoFields=[]</em><span class="sig-paren">)</span><a class="headerlink" href="#PyPenetrance.PyPenetrance" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a Python hybrid penetrance operator that passes genotype
at specified <em>loci</em>, values at specified information fields (if
requested), and a generation number to a user-defined function
<em>func</em>. Parameter <em>loci</em> can be a list of loci indexes, names,
list of chromosome position pairs, <code class="docutils literal notranslate"><span class="pre">ALL_AVAIL</span></code>, or a function
with optional parameter <code class="docutils literal notranslate"><span class="pre">pop</span></code> that will be called at each
ganeeration to determine indexes of loci. The return value will
be treated as  Individual penetrance.</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="class-pymlpenetrance">
<h2>class PyMlPenetrance<a class="headerlink" href="#class-pymlpenetrance" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="PyMlPenetrance">
<em class="property">class </em><code class="descname">PyMlPenetrance</code><a class="headerlink" href="#PyMlPenetrance" title="Permalink to this definition">¶</a></dt>
<dd><p>This penetrance operator is a multi-locus Python penetrance
operator that assigns penetrance values by combining locus and
genotype specific penetrance values. It differs from a
<a class="reference internal" href="#PyPenetrance" title="PyPenetrance"><code class="xref py py-class docutils literal notranslate"><span class="pre">PyPenetrance</span></code></a> in that the python function is responsible
for penetrance values values for each gentoype type at each locus,
which can potentially be random, and locus or gentoype-specific.</p>
<dl class="method">
<dt id="PyMlPenetrance.PyMlPenetrance">
<code class="descname">PyMlPenetrance</code><span class="sig-paren">(</span><em>func</em>, <em>mode=MULTIPLICATIVE</em>, <em>loci=ALL_AVAIL</em>, <em>ancGens=UNSPECIFIED</em>, <em>output=&quot;&quot;</em>, <em>begin=0</em>, <em>end=-1</em>, <em>step=1</em>, <em>at=[]</em>, <em>reps=ALL_AVAIL</em>, <em>subPops=ALL_AVAIL</em>, <em>infoFields=[]</em><span class="sig-paren">)</span><a class="headerlink" href="#PyMlPenetrance.PyMlPenetrance" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a penetrance operator that assigns individual affection
status according to penetrance values combined from locus-
specific penetrance values that are determined by a Python call-
back function. The callback function accepts parameter <em>loc</em>,
<em>alleles</em> (both optional) and returns location- or genotype-
specific penetrance values that can be constant or random. The
penetrance values for each genotype will be cached so the same
penetrance values will be assigned to genotypes with previously
assigned values. Note that a function that does not examine the
genotype naturally assumes a dominant model where genotypes with
one or two mutants have the same penetrance value. Because
genotypes at a locus are passed separately and in no particular
order, this function is also responsible for assigning
consistent fitness values for genotypes at the same locus (a
class is usually used). This operator currently ignores
chromosome types so unused alleles will be passed for loci on
sex or mitochondrial chromosomes. This operator also ignores the
phase of genotype so genotypes (a,b) and (b,a) are assumed to
have the same fitness effect.</p>
<p>Individual penetrance will be combined in <code class="docutils literal notranslate"><span class="pre">ADDITIVE</span></code>,
<code class="docutils literal notranslate"><span class="pre">MULTIPLICATIVE</span></code>, or <code class="docutils literal notranslate"><span class="pre">HETEROGENEITY</span></code> mode from penetrance
values of loci with at least one non-zero allele (See
<a class="reference internal" href="#MlPenetrance" title="MlPenetrance"><code class="xref py py-class docutils literal notranslate"><span class="pre">MlPenetrance</span></code></a> for details).</p>
</dd></dl>

</dd></dl>

</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">simuPOP</a></h1>








<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="userGuide_ch1.html">Front Matter</a></li>
<li class="toctree-l1"><a class="reference internal" href="userGuide_ch2.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="userGuide_ch3.html">Loading and running simuPOP</a></li>
<li class="toctree-l1"><a class="reference internal" href="userGuide_ch4.html">Individuals and Populations</a></li>
<li class="toctree-l1"><a class="reference internal" href="userGuide_ch5.html">simuPOP Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="userGuide_ch6.html">Evolving populations</a></li>
<li class="toctree-l1"><a class="reference internal" href="userGuide_ch7.html">Utility Modules</a></li>
<li class="toctree-l1"><a class="reference internal" href="userGuide_ch8.html">A real world example</a></li>
<li class="toctree-l1"><a class="reference internal" href="refManual_ch1.html">Front Matter</a></li>
<li class="toctree-l1"><a class="reference internal" href="refManual_ch2.html">simuPOP Components</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="refManual_ch3.html">Operator References</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="refManual_ch3_sec1.html">Base class for all operators</a></li>
<li class="toctree-l2"><a class="reference internal" href="refManual_ch3_sec2.html">Initialization</a></li>
<li class="toctree-l2"><a class="reference internal" href="refManual_ch3_sec3.html">Expression and Statements</a></li>
<li class="toctree-l2"><a class="reference internal" href="refManual_ch3_sec4.html">Demographic models</a></li>
<li class="toctree-l2"><a class="reference internal" href="refManual_ch3_sec5.html">Genotype transmitters</a></li>
<li class="toctree-l2"><a class="reference internal" href="refManual_ch3_sec6.html">Mutation</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Penetrance</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#class-basepenetrance">class BasePenetrance</a></li>
<li class="toctree-l3"><a class="reference internal" href="#class-mappenetrance">class MapPenetrance</a></li>
<li class="toctree-l3"><a class="reference internal" href="#class-mapenetrance">class MaPenetrance</a></li>
<li class="toctree-l3"><a class="reference internal" href="#class-mlpenetrance">class MlPenetrance</a></li>
<li class="toctree-l3"><a class="reference internal" href="#class-pypenetrance">class PyPenetrance</a></li>
<li class="toctree-l3"><a class="reference internal" href="#class-pymlpenetrance">class PyMlPenetrance</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="refManual_ch3_sec8.html">Quantitative Trait</a></li>
<li class="toctree-l2"><a class="reference internal" href="refManual_ch3_sec9.html">Natural selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="refManual_ch3_sec10.html">Tagging operators</a></li>
<li class="toctree-l2"><a class="reference internal" href="refManual_ch3_sec11.html">Statistics Calculation</a></li>
<li class="toctree-l2"><a class="reference internal" href="refManual_ch3_sec12.html">Conditional operators</a></li>
<li class="toctree-l2"><a class="reference internal" href="refManual_ch3_sec13.html">The Python operator</a></li>
<li class="toctree-l2"><a class="reference internal" href="refManual_ch3_sec14.html">Miscellaneous operators</a></li>
<li class="toctree-l2"><a class="reference internal" href="refManual_ch3_sec15.html">Function form of operators</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="refManual_ch4.html">Utility Modules</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="refManual.html">&lt;no title&gt;</a><ul>
  <li><a href="refManual_ch3.html">Operator References</a><ul>
      <li>Previous: <a href="refManual_ch3_sec6.html" title="previous chapter">Mutation</a></li>
      <li>Next: <a href="refManual_ch3_sec8.html" title="next chapter">Quantitative Trait</a></li>
  </ul></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2004-2019, Bo Peng.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.8.4</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/refManual_ch3_sec7.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>