
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Penetrance &#8212; simuPOP  documentation</title>
    <link rel="stylesheet" href="_static/simuPOP.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Quantitative trait" href="userGuide_ch5_sec8.html" />
    <link rel="prev" title="Mutation" href="userGuide_ch5_sec6.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="penetrance">
<h1>Penetrance<a class="headerlink" href="#penetrance" title="Permalink to this headline">¶</a></h1>
<p>Penetrance is the probability for an individual to be affected with a disease
conditioning on his or her genotype and other risk factors. A penetrance model
calculates such a probability for an individual and assign affection status
randomly according to this probability. For example, if an individual with
genotype <code class="docutils literal notranslate"><span class="pre">10</span></code> has probability 0.2 to be affected according to a penetrance
model, he or she will be affected with probability 0.2. Note that simuPOP
supports only one affection status. If there are multiple affection outcomes
involved, you can treat them as binary quantitative traits and use information
fields to store them.</p>
<p>A penetrance operator can be applied before or after mating, to assign affection
status to all individuals in the parental or offspring generation, respectively.
It can also be applied during mating and assign affection status to each
offspring. The latter could be used to assit natural selection through the
selection of offspring. You can also assign affection status to all individuals
in a population using the function form of a penetrance operator (e.g. function
<code class="docutils literal notranslate"><span class="pre">mapPenetrance</span></code> for operator <a class="reference internal" href="refManual_ch3_sec7.html#MapPenetrance" title="MapPenetrance"><code class="xref py py-class docutils literal notranslate"><span class="pre">MapPenetrance</span></code></a>). Compared the penetrance
operators that assign affection status to only the current generation, <strong>these
functions by default assign affection status to all ancestral generations as
well</strong>.</p>
<p>A penetrance operator usually do not store the penetrance values. However, if an
information field is given, penetrance values will be saved to this information
field before it is used to determine individual affection status.</p>
<div class="section" id="map-penetrance-model-operator-mappenetrance">
<h2>Map penetrance model (operator <a class="reference internal" href="refManual_ch3_sec7.html#MapPenetrance" title="MapPenetrance"><code class="xref py py-class docutils literal notranslate"><span class="pre">MapPenetrance</span></code></a>)<a class="headerlink" href="#map-penetrance-model-operator-mappenetrance" title="Permalink to this headline">¶</a></h2>
<p>A map penetrance opertor uses a Python dictionary to provide penetrance values
for each type of genotype. For example, Example <a class="reference internal" href="#mappenetrance"><span class="std std-ref">MapPenetrance</span></a> uses a dictionary with keys <code class="docutils literal notranslate"><span class="pre">(0,0)</span></code>, <code class="docutils literal notranslate"><span class="pre">(0,1)</span></code> and <code class="docutils literal notranslate"><span class="pre">(1,1)</span></code>
to specify penetrance for individuals with these genotypes at locus 0.</p>
<p id="mappenetrance"><strong>Example</strong>: <em>A penetrance model that uses pre-defined fitness value</em></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">simuPOP</span> <span class="k">as</span> <span class="nn">sim</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">sim</span><span class="o">.</span><span class="n">Population</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="mi">2000</span><span class="p">,</span> <span class="n">loci</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sim</span><span class="o">.</span><span class="n">initGenotype</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="n">freq</span><span class="o">=</span><span class="p">[</span><span class="o">.</span><span class="mi">2</span><span class="p">,</span> <span class="o">.</span><span class="mi">8</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sim</span><span class="o">.</span><span class="n">mapPenetrance</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="n">loci</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">penetrance</span><span class="o">=</span><span class="p">{(</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">):</span><span class="mi">0</span><span class="p">,</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">):</span><span class="o">.</span><span class="mi">2</span><span class="p">,</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">):</span><span class="o">.</span><span class="mi">3</span><span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sim</span><span class="o">.</span><span class="n">stat</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="n">genoFreq</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">numOfAffected</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="nb">vars</span><span class="o">=</span><span class="s1">&#39;genoNum&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># number of affected individuals</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span><span class="o">.</span><span class="n">dvars</span><span class="p">()</span><span class="o">.</span><span class="n">numOfAffected</span>
<span class="go">531</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># which should be roughly (#01 + #10) * 0.2 + #11 * 0.3</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">(</span><span class="n">pop</span><span class="o">.</span><span class="n">dvars</span><span class="p">()</span><span class="o">.</span><span class="n">genoNum</span><span class="p">[</span><span class="mi">0</span><span class="p">][(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">)]</span> <span class="o">+</span> <span class="n">pop</span><span class="o">.</span><span class="n">dvars</span><span class="p">()</span><span class="o">.</span><span class="n">genoNum</span><span class="p">[</span><span class="mi">0</span><span class="p">][(</span><span class="mi">1</span><span class="p">,</span><span class="mi">0</span><span class="p">)])</span> <span class="o">*</span> <span class="mf">0.2</span> \
<span class="gp">... </span><span class="o">+</span> <span class="n">pop</span><span class="o">.</span><span class="n">dvars</span><span class="p">()</span><span class="o">.</span><span class="n">genoNum</span><span class="p">[</span><span class="mi">0</span><span class="p">][(</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">)]</span> <span class="o">*</span> <span class="mf">0.3</span>
<span class="go">514.2</span>

<span class="go">now exiting runScriptInteractively...</span>
</pre></div>
</div>
<p><a class="reference external" href="MapPenetrance.py">Download MapPenetrance.py</a></p>
<p>The above example assumes that penetrance for individuals with genotypes
<code class="docutils literal notranslate"><span class="pre">(0,1)</span></code> and <code class="docutils literal notranslate"><span class="pre">(1,0)</span></code> are the same. This assumption is usually valid but can
be vialoated with impriting. In that case, you can specify fitness for both
types of genotypes. The underlying mechanism is that the <a class="reference internal" href="refManual_ch3_sec7.html#MapPenetrance" title="MapPenetrance"><code class="xref py py-class docutils literal notranslate"><span class="pre">MapPenetrance</span></code></a>
looks up a genotype in the dictionary first directly, and then without phase
information if a genotype is not found.</p>
<p>This operator supports haplodiploid populations and sex chromosomes. In these
cases, only valid alleles should be listed which can lead to dictionary keys
with different lengths. In addition, although less used because of potentially a
large number of keys, this operator can act on multiple loci. For example,</p>
<ul class="simple">
<li>keys <code class="docutils literal notranslate"><span class="pre">(a1,a2)</span></code> and <code class="docutils literal notranslate"><span class="pre">(a1,)</span></code> can be used to specify fitness values for
female and male individuals in a haplodiploid population, respectively</li>
<li>keys <code class="docutils literal notranslate"><span class="pre">(x1,x2)</span></code> and <code class="docutils literal notranslate"><span class="pre">(x1,)</span></code> can be used to specify fitness for female and
male individuals according to a locus on the X chromosome in a diploid
population, respectively. Similarly, keys <code class="docutils literal notranslate"><span class="pre">()</span></code> and <code class="docutils literal notranslate"><span class="pre">(y,)</span></code> for a locus on
chromosome Y.</li>
<li>keys <code class="docutils literal notranslate"><span class="pre">(a1,a2,b1,b2)</span></code> can be used to specify fitness values according to
genotype at two loci in a diploid population.</li>
</ul>
</div>
<div class="section" id="multi-allele-penetrance-model-operator-mapenetrance">
<h2>Multi-allele penetrance model (operator <a class="reference internal" href="refManual_ch3_sec7.html#MaPenetrance" title="MaPenetrance"><code class="xref py py-class docutils literal notranslate"><span class="pre">MaPenetrance</span></code></a>)<a class="headerlink" href="#multi-allele-penetrance-model-operator-mapenetrance" title="Permalink to this headline">¶</a></h2>
<p>A multi-allele penetrance model divides alleles into two groups, wildtype <em>A</em>
and mutants <em>a</em>, and treat alleles within each group as the same. The penetrance
model is therefore simplified to</p>
<ul class="simple">
<li>Two fitness values for genotype <span class="math notranslate nohighlight">\(A\)</span>, <span class="math notranslate nohighlight">\(a\)</span> in the haploid case</li>
<li>Three fitness values for genotype <em>AA</em>, <em>Aa</em> and <em>aa</em> in the diploid single
locus case. Genotype <em>Aa</em> and <em>aA</em> are assumed to have the same impact on
fitness.</li>
</ul>
<p>The default wildtype group contains allele 0 so the two allele groups are zero
and non-zero alleles. Example <a class="reference internal" href="#mapenetrance"><span class="std std-ref">MaPenetrance</span></a> demonstrates
the use of this operator.</p>
<p id="mapenetrance"><strong>Example</strong>: <em>A multi-allele penetrance model</em></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">simuPOP</span> <span class="k">as</span> <span class="nn">sim</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">sim</span><span class="o">.</span><span class="n">Population</span><span class="p">(</span><span class="mi">5000</span><span class="p">,</span> <span class="n">loci</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span><span class="o">.</span><span class="n">evolve</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">initOps</span><span class="o">=</span><span class="p">[</span>
<span class="gp">... </span>        <span class="n">sim</span><span class="o">.</span><span class="n">InitSex</span><span class="p">(),</span>
<span class="gp">... </span>        <span class="n">sim</span><span class="o">.</span><span class="n">InitGenotype</span><span class="p">(</span><span class="n">freq</span><span class="o">=</span><span class="p">[</span><span class="mf">0.9</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="mf">0.02</span><span class="p">]</span><span class="o">*</span><span class="mi">5</span><span class="p">)</span>
<span class="gp">... </span>    <span class="p">],</span>
<span class="gp">... </span>    <span class="n">matingScheme</span><span class="o">=</span><span class="n">sim</span><span class="o">.</span><span class="n">RandomMating</span><span class="p">(),</span>
<span class="gp">... </span>    <span class="n">postOps</span><span class="o">=</span><span class="p">[</span>
<span class="gp">... </span>        <span class="n">sim</span><span class="o">.</span><span class="n">MaPenetrance</span><span class="p">(</span><span class="n">loci</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">penetrance</span><span class="o">=</span><span class="p">(</span><span class="mf">0.01</span><span class="p">,</span> <span class="mf">0.2</span><span class="p">,</span> <span class="mf">0.3</span><span class="p">)),</span>
<span class="gp">... </span>        <span class="n">sim</span><span class="o">.</span><span class="n">Stat</span><span class="p">(</span><span class="n">numOfAffected</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="nb">vars</span><span class="o">=</span><span class="s1">&#39;propOfAffected&#39;</span><span class="p">),</span>
<span class="gp">... </span>        <span class="n">sim</span><span class="o">.</span><span class="n">PyEval</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;&#39;Gen: </span><span class="si">%d</span><span class="s2"> Prevalence: </span><span class="si">%.1f%%</span><span class="s2">\n&#39; % (gen, propOfAffected*100)&quot;</span><span class="p">),</span>
<span class="gp">... </span>    <span class="p">],</span>
<span class="gp">... </span>    <span class="n">gen</span> <span class="o">=</span> <span class="mi">5</span>
<span class="gp">... </span><span class="p">)</span>
<span class="go">Gen: 0 Prevalence: 4.4%</span>
<span class="go">Gen: 1 Prevalence: 4.4%</span>
<span class="go">Gen: 2 Prevalence: 4.7%</span>
<span class="go">Gen: 3 Prevalence: 4.4%</span>
<span class="go">Gen: 4 Prevalence: 4.3%</span>
<span class="go">5</span>

<span class="go">now exiting runScriptInteractively...</span>
</pre></div>
</div>
<p><a class="reference external" href="MaPenetrance.py">Download MaPenetrance.py</a></p>
<p>Operator <a class="reference internal" href="refManual_ch3_sec7.html#MaPenetrance" title="MaPenetrance"><code class="xref py py-class docutils literal notranslate"><span class="pre">MaPenetrance</span></code></a> also supports multiple loci by specifying fitness
values for all combination of genotype at specified loci. In the case of two
loci, this operator requires</p>
<ul class="simple">
<li>Four fitness values for genotype <code class="docutils literal notranslate"><span class="pre">AB</span></code>, <code class="docutils literal notranslate"><span class="pre">Ab</span></code>, <code class="docutils literal notranslate"><span class="pre">aB</span></code> and <code class="docutils literal notranslate"><span class="pre">ab</span></code> in the
haploid case,</li>
<li>Nine fitness values for genotype <code class="docutils literal notranslate"><span class="pre">AABB</span></code>, <code class="docutils literal notranslate"><span class="pre">AABb</span></code>, <code class="docutils literal notranslate"><span class="pre">AAbb</span></code>, <code class="docutils literal notranslate"><span class="pre">AaBB</span></code>,
<code class="docutils literal notranslate"><span class="pre">AaBb</span></code>, <code class="docutils literal notranslate"><span class="pre">Aabb</span></code>, <code class="docutils literal notranslate"><span class="pre">aaBB</span></code>, <code class="docutils literal notranslate"><span class="pre">aaBb</span></code>, and <code class="docutils literal notranslate"><span class="pre">aabb</span></code> in the haploid case.</li>
</ul>
<p>In general, <span class="math notranslate nohighlight">\(2^{n}\)</span> values are needed for haploid populations and
<span class="math notranslate nohighlight">\(3^{n}\)</span> values are needed for diploid populations where <span class="math notranslate nohighlight">\(n\)</span> is the
number of loci. This operator does not yet support haplodiploid populations and
sex chromosomes.</p>
</div>
<div class="section" id="multi-loci-penetrance-model-operator-mlpenetrance">
<h2>Multi-loci penetrance model (operator <a class="reference internal" href="refManual_ch3_sec7.html#MlPenetrance" title="MlPenetrance"><code class="xref py py-class docutils literal notranslate"><span class="pre">MlPenetrance</span></code></a>)<a class="headerlink" href="#multi-loci-penetrance-model-operator-mlpenetrance" title="Permalink to this headline">¶</a></h2>
<p>Although an individual’s affection status can be affected by several factors,
each of which can be modeled individually, <strong>only one penetrance value is used
to determine a person’s affection status</strong> and we have to use a multi-locus
penetrance model to combine single-locus models.</p>
<p>This multi-loci penetrance model applies several penetrance models to each
Individual and computes an overall penetrance value from the penetrance values
provided by these operators. Although this selector is designed to obtain multi-
loci penetrance values from several single-locus penetrance models, any
penetrance operator, including those obtain their penetrance values from
multiple disease predisposing loci, can be used in this operator. This operator
uses parameter <code class="docutils literal notranslate"><span class="pre">mode</span></code> to control how Individual penetrance values are
combined. More specifically, if <span class="math notranslate nohighlight">\(f_{i}\)</span> are penetrance values obtained
from individual selectors, this selector returns</p>
<ul class="simple">
<li><span class="math notranslate nohighlight">\(\Pi_{i}f_{i}\)</span> if <code class="docutils literal notranslate"><span class="pre">mode=MULTIPLICATIVE</span></code>, and</li>
<li><span class="math notranslate nohighlight">\(\sum_{i}f_{i}\)</span> if <code class="docutils literal notranslate"><span class="pre">mode=ADDITIVE</span></code>, and</li>
<li><span class="math notranslate nohighlight">\(1-\Pi_{i}\left(1-f_{i}\right)\)</span> if <code class="docutils literal notranslate"><span class="pre">mode=HETEROGENEITY</span></code></li>
</ul>
<p>0 or 1 will be returned if the returned fitness value is out of range of
<code class="docutils literal notranslate"><span class="pre">[0,1]</span></code>.</p>
<p>Example <a class="reference internal" href="#mlpenetrance"><span class="std std-ref">MlPenetrance</span></a> demonstrates the use of this operator
using an multiplicative multi-locus model over three additive single-locus
models at three diesease predisposing loci.</p>
<p id="mlpenetrance"><strong>Example</strong>: <em>A multi-loci penetrance model</em></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">simuPOP</span> <span class="k">as</span> <span class="nn">sim</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">sim</span><span class="o">.</span><span class="n">Population</span><span class="p">(</span><span class="mi">5000</span><span class="p">,</span> <span class="n">loci</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sim</span><span class="o">.</span><span class="n">initGenotype</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="n">freq</span><span class="o">=</span><span class="p">[</span><span class="mf">0.2</span><span class="p">]</span><span class="o">*</span><span class="mi">5</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># the multi-loci penetrance</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sim</span><span class="o">.</span><span class="n">mlPenetrance</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="n">sim</span><span class="o">.</span><span class="n">MULTIPLICATIVE</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">ops</span> <span class="o">=</span> <span class="p">[</span><span class="n">sim</span><span class="o">.</span><span class="n">MaPenetrance</span><span class="p">(</span><span class="n">loci</span><span class="o">=</span><span class="n">loc</span><span class="p">,</span>
<span class="gp">... </span>        <span class="n">penetrance</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mf">0.3</span><span class="p">,</span> <span class="mf">0.6</span><span class="p">])</span> <span class="k">for</span> <span class="n">loc</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">3</span><span class="p">)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># count the number of affected individuals.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sim</span><span class="o">.</span><span class="n">stat</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="n">numOfAffected</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span><span class="o">.</span><span class="n">dvars</span><span class="p">()</span><span class="o">.</span><span class="n">numOfAffected</span>
<span class="go">542</span>

<span class="go">now exiting runScriptInteractively...</span>
</pre></div>
</div>
<p><a class="reference external" href="MlPenetrance.py">Download MlPenetrance.py</a></p>
</div>
<div class="section" id="hybrid-penetrance-model-operator-pypenetrance">
<h2>Hybrid penetrance model (operator <a class="reference internal" href="refManual_ch3_sec7.html#PyPenetrance" title="PyPenetrance"><code class="xref py py-class docutils literal notranslate"><span class="pre">PyPenetrance</span></code></a>)<a class="headerlink" href="#hybrid-penetrance-model-operator-pypenetrance" title="Permalink to this headline">¶</a></h2>
<p>When your selection model involves multiple interacting genetic and
environmental factors, it might be easier to calculate a penetrance value
explicitly using a Python function. A hybrid penetrance operator can be used for
this purpose. If your penetrance model depends solely on genotype, you can
define a function such as</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">pfunc</span><span class="p">(</span><span class="n">geno</span><span class="p">):</span>
    <span class="c1"># calculate penetrance according to genotype at specified loci</span>
    <span class="c1"># in the order of A1,A2,B1,B2,C1,C2 for loci A,B,C (for diploid)</span>
    <span class="k">return</span> <span class="n">val</span>
</pre></div>
</div>
<p>and use this function in an operator <a class="reference internal" href="refManual_ch3_sec9.html#PySelector" title="PySelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">PySelector</span></code></a>(<code class="docutils literal notranslate"><span class="pre">func=pfunc,</span>
<span class="pre">loci=loci</span></code>). If your penetrance model depends on genotype as well as some
information fields, you can define a function in the form of</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">pfunc</span><span class="p">(</span><span class="n">geno</span><span class="p">,</span> <span class="n">fields</span><span class="p">):</span>
    <span class="c1"># calculate penetrance according to genotype at specified loci</span>
    <span class="c1"># and values at specified informaton fields.</span>
    <span class="k">return</span> <span class="n">val</span>
</pre></div>
</div>
<p>and use this function in an operator <a class="reference internal" href="refManual_ch3_sec9.html#PySelector" title="PySelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">PySelector</span></code></a>(<code class="docutils literal notranslate"><span class="pre">func=pfunc,</span>
<span class="pre">loci=loci,</span> <span class="pre">paramFields=fields</span></code>). If the function you provide accepts three
arguments, <a class="reference internal" href="refManual_ch3_sec7.html#PyPenetrance" title="PyPenetrance"><code class="xref py py-class docutils literal notranslate"><span class="pre">PyPenetrance</span></code></a> will pass generation number as the third
argument so that you could implement generation-specific penetrance models (e.g.
<code class="docutils literal notranslate"><span class="pre">pfunc(geno,</span> <span class="pre">fields,</span> <span class="pre">gen)</span></code>).</p>
<p>When a <a class="reference internal" href="refManual_ch3_sec7.html#PyPenetrance" title="PyPenetrance"><code class="xref py py-class docutils literal notranslate"><span class="pre">PyPenetrance</span></code></a> operator is used to calculate penetrance for an
individual, it will collect his or her genotype at specified loci, optional
values at specified information fields, and the generation number to a user-
specified Python function, and take its return value as penetrance. As you can
imagine, the incorporation of information fields and generation number allow the
implementation of very complex penetrance scenarios such as gene environment
interaction and varying selection pressures. Note that this operator does not
pass sex and affection status to the user-defined function. If your selection
model is sex-dependent, you can define an information field <code class="docutils literal notranslate"><span class="pre">sex</span></code>, synchronize
its value with individual sex (e.g. using operator <a class="reference internal" href="refManual_ch3_sec3.html#InfoExec" title="InfoExec"><code class="xref py py-class docutils literal notranslate"><span class="pre">InfoExec</span></code></a><code class="docutils literal notranslate"><span class="pre">('sex=ind.sex()',</span> <span class="pre">exposeInd='ind'</span></code>) and pass this information to the user-
defined function (<a class="reference internal" href="refManual_ch3_sec9.html#PySelector" title="PySelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">PySelector</span></code></a>(<code class="docutils literal notranslate"><span class="pre">func=func,</span> <span class="pre">paramFields='sex'</span></code>)).</p>
<p>Example <a class="reference internal" href="userGuide_ch5_sec9.html#pyselector"><span class="std std-ref">PySelector</span></a> demonstrates how to use a
<a class="reference internal" href="refManual_ch3_sec7.html#PyPenetrance" title="PyPenetrance"><code class="xref py py-class docutils literal notranslate"><span class="pre">PyPenetrance</span></code></a> to specify penetrance values according to a fitness table
and the smoking status of each individual. In this example, Individual risk is
doubled when he or she smokes. The disease prevalence is therefore much higher
in smokers than in non-smokers.</p>
<p id="pypenetrance"><strong>Example</strong>: <em>A hybrid penetrance model</em></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">simuPOP</span> <span class="k">as</span> <span class="nn">sim</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">random</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">sim</span><span class="o">.</span><span class="n">Population</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="mi">2000</span><span class="p">,</span> <span class="n">loci</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">*</span><span class="mi">2</span><span class="p">,</span> <span class="n">infoFields</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;p&#39;</span><span class="p">,</span> <span class="s1">&#39;smoking&#39;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span><span class="o">.</span><span class="n">setVirtualSplitter</span><span class="p">(</span><span class="n">sim</span><span class="o">.</span><span class="n">InfoSplitter</span><span class="p">(</span><span class="n">field</span><span class="o">=</span><span class="s1">&#39;smoking&#39;</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">]))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># the second parameter gen can be used for varying selection pressure</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">def</span> <span class="nf">penet</span><span class="p">(</span><span class="n">geno</span><span class="p">,</span> <span class="n">smoking</span><span class="p">):</span>
<span class="gp">... </span>    <span class="c1">#     BB     Bb      bb</span>
<span class="gp">... </span>    <span class="c1"># AA  0.01   0.01    0.01</span>
<span class="gp">... </span>    <span class="c1"># Aa  0.01   0.03    0.03</span>
<span class="gp">... </span>    <span class="c1"># aa  0.01   0.03    0.05</span>
<span class="gp">... </span>    <span class="c1">#</span>
<span class="gp">... </span>    <span class="c1"># geno is (A1 A2 B1 B2)</span>
<span class="gp">... </span>    <span class="k">if</span> <span class="n">geno</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">geno</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span> <span class="ow">and</span> <span class="n">geno</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">+</span> <span class="n">geno</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
<span class="gp">... </span>        <span class="n">v</span> <span class="o">=</span> <span class="mf">0.03</span>   <span class="c1"># case of AaBb</span>
<span class="gp">... </span>    <span class="k">elif</span> <span class="n">geno</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">geno</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="mi">2</span> <span class="ow">and</span> <span class="n">geno</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">+</span> <span class="n">geno</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
<span class="gp">... </span>        <span class="n">v</span> <span class="o">=</span> <span class="mf">0.03</span>   <span class="c1"># case of aaBb</span>
<span class="gp">... </span>    <span class="k">elif</span> <span class="n">geno</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">geno</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span><span class="mi">2</span> <span class="ow">and</span> <span class="n">geno</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">+</span> <span class="n">geno</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>
<span class="gp">... </span>        <span class="n">v</span> <span class="o">=</span> <span class="mf">0.05</span>   <span class="c1"># case of aabb</span>
<span class="gp">... </span>    <span class="k">else</span><span class="p">:</span>
<span class="gp">... </span>        <span class="n">v</span> <span class="o">=</span> <span class="mf">0.01</span>   <span class="c1"># other cases</span>
<span class="gp">... </span>    <span class="k">if</span> <span class="n">smoking</span><span class="p">:</span>
<span class="gp">... </span>        <span class="k">return</span> <span class="n">v</span> <span class="o">*</span> <span class="mi">2</span>
<span class="gp">... </span>    <span class="k">else</span><span class="p">:</span>
<span class="gp">... </span>        <span class="k">return</span> <span class="n">v</span>
<span class="gp">...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span><span class="o">.</span><span class="n">evolve</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">initOps</span><span class="o">=</span><span class="p">[</span>
<span class="gp">... </span>        <span class="n">sim</span><span class="o">.</span><span class="n">InitSex</span><span class="p">(),</span>
<span class="gp">... </span>        <span class="n">sim</span><span class="o">.</span><span class="n">InitGenotype</span><span class="p">(</span><span class="n">freq</span><span class="o">=</span><span class="p">[</span><span class="o">.</span><span class="mi">5</span><span class="p">,</span> <span class="o">.</span><span class="mi">5</span><span class="p">]),</span>
<span class="gp">... </span>        <span class="n">sim</span><span class="o">.</span><span class="n">PyOutput</span><span class="p">(</span><span class="s1">&#39;Calculate prevalence in smoker and non-smokers</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">),</span>
<span class="gp">... </span>    <span class="p">],</span>
<span class="gp">... </span>    <span class="n">matingScheme</span><span class="o">=</span><span class="n">sim</span><span class="o">.</span><span class="n">RandomMating</span><span class="p">(),</span>
<span class="gp">... </span>    <span class="n">postOps</span><span class="o">=</span><span class="p">[</span>
<span class="gp">... </span>        <span class="c1"># set smoking status randomly</span>
<span class="gp">... </span>        <span class="n">sim</span><span class="o">.</span><span class="n">InitInfo</span><span class="p">(</span><span class="k">lambda</span> <span class="p">:</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span> <span class="n">infoFields</span><span class="o">=</span><span class="s1">&#39;smoking&#39;</span><span class="p">),</span>
<span class="gp">... </span>        <span class="c1"># assign affection status</span>
<span class="gp">... </span>        <span class="n">sim</span><span class="o">.</span><span class="n">PyPenetrance</span><span class="p">(</span><span class="n">loci</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">func</span><span class="o">=</span><span class="n">penet</span><span class="p">),</span>
<span class="gp">... </span>        <span class="n">sim</span><span class="o">.</span><span class="n">Stat</span><span class="p">(</span><span class="n">numOfAffected</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">subPops</span><span class="o">=</span><span class="p">[(</span><span class="mi">0</span><span class="p">,</span> <span class="n">sim</span><span class="o">.</span><span class="n">ALL_AVAIL</span><span class="p">)],</span>
<span class="gp">... </span>            <span class="nb">vars</span><span class="o">=</span><span class="s1">&#39;propOfAffected_sp&#39;</span><span class="p">,</span> <span class="n">step</span><span class="o">=</span><span class="mi">20</span><span class="p">),</span>
<span class="gp">... </span>        <span class="n">sim</span><span class="o">.</span><span class="n">PyEval</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;&#39;Non-smoker: </span><span class="si">%.2f%%</span><span class="s2">\tSmoker: </span><span class="si">%.2f%%</span><span class="s2">\n&#39; % &quot;</span>
<span class="gp">... </span>            <span class="s2">&quot;(subPop[(0,0)][&#39;propOfAffected&#39;]*100, subPop[(0,1)][&#39;propOfAffected&#39;]*100)&quot;</span><span class="p">,</span>
<span class="gp">... </span>            <span class="n">step</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
<span class="gp">... </span>    <span class="p">],</span>
<span class="gp">... </span>    <span class="n">gen</span> <span class="o">=</span> <span class="mi">50</span>
<span class="gp">... </span><span class="p">)</span>
<span class="go">Calculate prevalence in smoker and non-smokers</span>
<span class="go">Non-smoker: 2.24%    Smoker: 4.52%</span>
<span class="go">Non-smoker: 2.29%    Smoker: 3.61%</span>
<span class="go">Non-smoker: 1.85%    Smoker: 3.80%</span>
<span class="go">50</span>
<span class="go">&gt;&gt;&gt;</span>

<span class="go">now exiting runScriptInteractively...</span>
</pre></div>
</div>
<p><a class="reference external" href="PyPenetrance.py">Download PyPenetrance.py</a></p>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">simuPOP</a></h1>








<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="userGuide_ch1.html">Front Matter</a></li>
<li class="toctree-l1"><a class="reference internal" href="userGuide_ch2.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="userGuide_ch3.html">Loading and running simuPOP</a></li>
<li class="toctree-l1"><a class="reference internal" href="userGuide_ch4.html">Individuals and Populations</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="userGuide_ch5.html">simuPOP Operators</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch5_sec1.html">Introduction to operators</a></li>
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch5_sec2.html">Initialization</a></li>
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch5_sec3.html">Expressions and statements</a></li>
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch5_sec4.html">Demographic changes</a></li>
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch5_sec5.html">Genotype transmitters</a></li>
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch5_sec6.html">Mutation</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Penetrance</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#map-penetrance-model-operator-mappenetrance">Map penetrance model (operator <code class="docutils literal notranslate"><span class="pre">MapPenetrance</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#multi-allele-penetrance-model-operator-mapenetrance">Multi-allele penetrance model (operator <code class="docutils literal notranslate"><span class="pre">MaPenetrance</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#multi-loci-penetrance-model-operator-mlpenetrance">Multi-loci penetrance model (operator <code class="docutils literal notranslate"><span class="pre">MlPenetrance</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#hybrid-penetrance-model-operator-pypenetrance">Hybrid penetrance model (operator <code class="docutils literal notranslate"><span class="pre">PyPenetrance</span></code>)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch5_sec8.html">Quantitative trait</a></li>
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch5_sec9.html">Natural Selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch5_sec10.html">Tagging operators</a></li>
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch5_sec11.html">Statistics calculation (operator <code class="docutils literal notranslate"><span class="pre">Stat</span></code>)</a></li>
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch5_sec12.html">Conditional operators</a></li>
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch5_sec13.html">Miscellaneous operators</a></li>
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch5_sec14.html">Hybrid and Python operators</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="userGuide_ch6.html">Evolving populations</a></li>
<li class="toctree-l1"><a class="reference internal" href="userGuide_ch7.html">Utility Modules</a></li>
<li class="toctree-l1"><a class="reference internal" href="userGuide_ch8.html">A real world example</a></li>
<li class="toctree-l1"><a class="reference internal" href="refManual_ch1.html">Front Matter</a></li>
<li class="toctree-l1"><a class="reference internal" href="refManual_ch2.html">simuPOP Components</a></li>
<li class="toctree-l1"><a class="reference internal" href="refManual_ch3.html">Operator References</a></li>
<li class="toctree-l1"><a class="reference internal" href="refManual_ch4.html">Utility Modules</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="userGuide.html">&lt;no title&gt;</a><ul>
  <li><a href="userGuide_ch5.html">simuPOP Operators</a><ul>
      <li>Previous: <a href="userGuide_ch5_sec6.html" title="previous chapter">Mutation</a></li>
      <li>Next: <a href="userGuide_ch5_sec8.html" title="next chapter">Quantitative trait</a></li>
  </ul></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2004-2019, Bo Peng.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.8.4</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/userGuide_ch5_sec7.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>