
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Mutation &#8212; simuPOP  documentation</title>
    <link rel="stylesheet" href="_static/simuPOP.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Penetrance" href="refManual_ch3_sec7.html" />
    <link rel="prev" title="Genotype transmitters" href="refManual_ch3_sec5.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="mutation">
<h1>Mutation<a class="headerlink" href="#mutation" title="Permalink to this headline">¶</a></h1>
<div class="section" id="class-basemutator">
<span id="index-0"></span><h2>class BaseMutator<a class="headerlink" href="#class-basemutator" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="BaseMutator">
<em class="property">class </em><code class="descname">BaseMutator</code><a class="headerlink" href="#BaseMutator" title="Permalink to this definition">¶</a></dt>
<dd><p>Class <code class="docutils literal notranslate"><span class="pre">mutator</span></code> is the base class of all mutators. It handles all
the work of picking an allele at specified loci from certain
(virtual) subpopulation with certain probability, and calling a
derived mutator to mutate the allele. Alleles can be changed before
and after mutation if existing allele numbers do not match those of
a mutation model.</p>
<dl class="method">
<dt id="BaseMutator.BaseMutator">
<code class="descname">BaseMutator</code><span class="sig-paren">(</span><em>rates=[], loci=ALL_AVAIL, mapIn=[], mapOut=[], context=0, output=&quot;&quot;, begin=0, end=-1, step=1, at=[], reps=ALL_AVAIL, subPops=ALL_AVAIL, infoFields=[&quot;ind_id&quot;], lineageMode=FROM_INFO</em><span class="sig-paren">)</span><a class="headerlink" href="#BaseMutator.BaseMutator" title="Permalink to this definition">¶</a></dt>
<dd><p>A mutator mutates alleles from one state to another with given
probability. This base mutator does not perform any mutation but
it defines common behaviors of all mutators.</p>
<p>By default, a mutator mutates all alleles in all populations of
a simulator at all generations. A number of parameters can be
used to restrict mutations to certain generations (parameters
<em>begin</em>, <em>end</em>, <em>step</em> and <em>at</em>), replicate populations
(parameter <em>rep</em>), (virtual) subpopulations (parameter
<em>subPops</em>) and loci (parameter <em>loci</em>). Parameter <em>loci</em> can be
a list of loci indexes, names, list of chromosome position
pairs, <code class="docutils literal notranslate"><span class="pre">ALL_AVAIL</span></code>, or a function with optional parameter
<code class="docutils literal notranslate"><span class="pre">pop</span></code> that will be called at each ganeeration to determine
indexes of loci. Please refer to class <a class="reference internal" href="refManual_ch3_sec1.html#BaseOperator" title="BaseOperator"><code class="xref py py-class docutils literal notranslate"><span class="pre">BaseOperator</span></code></a> for
a detailed explanation of these parameters.</p>
<p>Parameter <em>rate</em> or its equivalence specifies the probability
that a a mutation event happens. The exact form and meaning of
<em>rate</em> is mutator-specific. If a single rate is specified, it
will be applied to all <em>loci</em>. If a list of mutation rates are
given, they will be applied to each locus specified in parameter
<em>loci</em>. Note that not all mutators allow specification of
multiple mutation rate, especially when the mutation rate itself
is a list or matrix.</p>
<p>Alleles at a locus are non-negative numbers 0, 1, … up to the
maximum allowed allele for the loaded module (1 for binary, 255
for short and 65535 for long modules). Whereas some general
mutation models treat alleles as numbers, other models assume
specific interpretation of alleles. For example, an
<a class="reference internal" href="#AcgtMutator" title="AcgtMutator"><code class="xref py py-class docutils literal notranslate"><span class="pre">AcgtMutator</span></code></a> assumes alleles <code class="docutils literal notranslate"><span class="pre">0</span></code>, <code class="docutils literal notranslate"><span class="pre">1</span></code>, <code class="docutils literal notranslate"><span class="pre">2</span></code> and
<code class="docutils literal notranslate"><span class="pre">3</span></code> as nucleotides <code class="docutils literal notranslate"><span class="pre">A</span></code>, <code class="docutils literal notranslate"><span class="pre">C</span></code>, <code class="docutils literal notranslate"><span class="pre">G</span></code> and <code class="docutils literal notranslate"><span class="pre">T</span></code>. Using a
mutator that is incompatible with your simulation will certainly
yield erroneous results.</p>
<p>If your simulation assumes different alleles with a mutation
model, you can map an allele to the allele used in the model and
map the mutated allele back. This is achieved using a <em>mapIn</em>
list with its <code class="docutils literal notranslate"><span class="pre">i-th</span></code> item being the corresponding allele of
real allele <code class="docutils literal notranslate"><span class="pre">i</span></code>, and a <em>mapOut</em> list with its <em>i-th</em> item
being the real allele of allele <code class="docutils literal notranslate"><span class="pre">i</span></code> assumed in the model. For
example <code class="docutils literal notranslate"><span class="pre">mapIn=[0,</span> <span class="pre">0,</span> <span class="pre">1]</span></code> and <code class="docutils literal notranslate"><span class="pre">mapOut=[1,</span> <span class="pre">2]</span></code> would allow
the use of a <a class="reference internal" href="#SNPMutator" title="SNPMutator"><code class="xref py py-class docutils literal notranslate"><span class="pre">SNPMutator</span></code></a> to mutate between alleles 1 and
2, instead of 0 and 1. Parameters <em>mapIn</em> and <em>mapOut</em> also
accept a user-defined Python function that returns a
corresponding allele for a given allele. This allows easier
mapping between a large number of alleles and advanced models
such as random emission of alleles.</p>
<p>If a valid information field is specified for parameter
<em>infoFields</em> (default to <code class="docutils literal notranslate"><span class="pre">ind_id</span></code>) for modules with lineage
allele type, the lineage of the mutated alleles will be the ID
(stored in the first field of <em>infoFields</em>) of individuals that
harbor the mutated alleles if <em>lineageMode</em> is set to
<code class="docutils literal notranslate"><span class="pre">FROM_INFO</span></code> (default). If <em>lineageMode</em> is set to
<code class="docutils literal notranslate"><span class="pre">FROM_INFO_SIGNED</span></code>, the IDs will be assigned a sign depending
on the ploidy the mutation happens (1 for ploidy 0, -1 for
ploidy 1, etc). The lineage information will be transmitted
along with the alleles so this feature allows you to track the
source of mutants during evolution.A</p>
<p>A mutator by default does not produce any output. However, if an
non-empty output is specified, the operator will output
generation number, locus, ploidy, original allele, mutant, and
values of all information field specified by parameter
<code class="docutils literal notranslate"><span class="pre">infoFields</span></code> (e.g. individual ID if <code class="docutils literal notranslate"><span class="pre">ind_id</span></code> is specified).</p>
<p>Some mutation models are context dependent. Namely, how an
allele mutates will depend on its adjecent alleles. Whereas most
simuPOP mutators are context independent, some of them accept a
parameter <em>context</em> which is the number of alleles to the left
and right of the mutated allele. For example <em>context=1</em> will
make the alleles to the immediate left and right to a mutated
allele available to a mutator. These alleles will be mapped in
if parameter <em>mapIn</em> is defined. How exactly a mutator makes use
of these information is mutator dependent.</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="class-matrixmutator">
<h2>class MatrixMutator<a class="headerlink" href="#class-matrixmutator" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="MatrixMutator">
<em class="property">class </em><code class="descname">MatrixMutator</code><a class="headerlink" href="#MatrixMutator" title="Permalink to this definition">¶</a></dt>
<dd><p>A matrix mutator mutates alleles <code class="docutils literal notranslate"><span class="pre">0</span></code>, <code class="docutils literal notranslate"><span class="pre">1</span></code>, …, <code class="docutils literal notranslate"><span class="pre">n-1</span></code> using a
<code class="docutils literal notranslate"><span class="pre">n</span></code> by <code class="docutils literal notranslate"><span class="pre">n</span></code> matrix, which specifies the probability at which
each allele mutates to another. Conceptually speaking, this mutator
goes through all mutable allele and mutate it to another state
according to probabilities in the corresponding row of the rate
matrix. Only one mutation rate matrix can be specified which will
be used for all specified loci. #</p>
<dl class="method">
<dt id="MatrixMutator.MatrixMutator">
<code class="descname">MatrixMutator</code><span class="sig-paren">(</span><em>rate, loci=ALL_AVAIL, mapIn=[], mapOut=[], output=&quot;&quot;, begin=0, end=-1, step=1, at=[], reps=ALL_AVAIL, subPops=ALL_AVAIL, infoFields=[&quot;ind_id&quot;], lineageMode=FROM_INFO</em><span class="sig-paren">)</span><a class="headerlink" href="#MatrixMutator.MatrixMutator" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a mutator that mutates alleles <code class="docutils literal notranslate"><span class="pre">0</span></code>, <code class="docutils literal notranslate"><span class="pre">1</span></code>, …, <code class="docutils literal notranslate"><span class="pre">n-1</span></code>
using a <code class="docutils literal notranslate"><span class="pre">n</span></code> by <code class="docutils literal notranslate"><span class="pre">n</span></code> matrix <code class="docutils literal notranslate"><span class="pre">rate</span></code>. Item <code class="docutils literal notranslate"><span class="pre">(i,j)</span></code> of this
matrix specifies the probability at which allele <em>i</em> mutates to
allele <em>j</em>. Diagnal items <code class="docutils literal notranslate"><span class="pre">(i,</span> <span class="pre">i)</span></code> are ignored because they
are automatically determined by other probabilities. Only one
mutation rate matrix can be specified which will be used for all
loci in the applied population, or loci specified by parameter
<em>loci</em>. If alleles other than <code class="docutils literal notranslate"><span class="pre">0</span></code>, <code class="docutils literal notranslate"><span class="pre">1</span></code>, …, <code class="docutils literal notranslate"><span class="pre">n-1</span></code> exist
in the population, they will not be mutated although a warning
message will be given if debugging code <code class="docutils literal notranslate"><span class="pre">DBG_WARNING</span></code> is
turned on. Please refer to classes <code class="docutils literal notranslate"><span class="pre">mutator</span></code> and
<a class="reference internal" href="refManual_ch3_sec1.html#BaseOperator" title="BaseOperator"><code class="xref py py-class docutils literal notranslate"><span class="pre">BaseOperator</span></code></a> for detailed explanation of other
parameters.</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="class-kallelemutator">
<h2>class KAlleleMutator<a class="headerlink" href="#class-kallelemutator" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="KAlleleMutator">
<em class="property">class </em><code class="descname">KAlleleMutator</code><a class="headerlink" href="#KAlleleMutator" title="Permalink to this definition">¶</a></dt>
<dd><p>This mutator implements a <em>k-allele</em> mutation model that assumes
<em>k</em> allelic states (alleles 0, 1, 2, …, <em>k-1</em>) at each locus.
When a mutation event happens, it mutates an allele to any other
states with equal probability.</p>
<dl class="method">
<dt id="KAlleleMutator.KAlleleMutator">
<code class="descname">KAlleleMutator</code><span class="sig-paren">(</span><em>k, rates=[], loci=ALL_AVAIL, mapIn=[], mapOut=[], output=&quot;&quot;, begin=0, end=-1, step=1, at=[], reps=ALL_AVAIL, subPops=ALL_AVAIL, infoFields=[&quot;ind_id&quot;], lineageMode=FROM_INFO</em><span class="sig-paren">)</span><a class="headerlink" href="#KAlleleMutator.KAlleleMutator" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a k-allele mutator that mutates alleles to one of the
other <code class="docutils literal notranslate"><span class="pre">k-1</span></code> alleles with equal probability. This mutator by
default applies to all loci unless parameter <em>loci</em> is
specified. A single mutation rate will be used for all loci if a
single value of parameter <em>rates</em> is given. Otherwise, a list of
mutation rates can be specified for each locus in parameter
<em>loci</em>. If the mutated allele is larger than or equal to <code class="docutils literal notranslate"><span class="pre">k</span></code>,
it will not be mutated. A warning message will be displayed if
debugging code <code class="docutils literal notranslate"><span class="pre">DBG_WARNING</span></code> is turned on. Please refer to
classes <code class="docutils literal notranslate"><span class="pre">mutator</span></code> and <a class="reference internal" href="refManual_ch3_sec1.html#BaseOperator" title="BaseOperator"><code class="xref py py-class docutils literal notranslate"><span class="pre">BaseOperator</span></code></a> for descriptions
of other parameters.</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="class-stepwisemutator">
<h2>class StepwiseMutator<a class="headerlink" href="#class-stepwisemutator" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="StepwiseMutator">
<em class="property">class </em><code class="descname">StepwiseMutator</code><a class="headerlink" href="#StepwiseMutator" title="Permalink to this definition">¶</a></dt>
<dd><p>A stepwise mutation model treats alleles at a locus as the number
of tandem repeats of microsatellite or minisatellite markers. When
a mutation event happens, the number of repeats (allele) either
increase or decrease. A standard stepwise mutation model increases
of decreases an allele by 1 with equal probability. More complex
models (generalized stepwise mutation model) are also allowed. Note
that an allele cannot be mutated beyond boundaries (0 and maximum
allowed allele).</p>
<dl class="method">
<dt id="StepwiseMutator.StepwiseMutator">
<code class="descname">StepwiseMutator</code><span class="sig-paren">(</span><em>rates=[], loci=ALL_AVAIL, incProb=0.5, maxAllele=0, mutStep=[], mapIn=[], mapOut=[], output=&quot;&quot;, begin=0, end=-1, step=1, at=[], reps=ALL_AVAIL, subPops=ALL_AVAIL, infoFields=[&quot;ind_id&quot;], lineageMode=FROM_INFO</em><span class="sig-paren">)</span><a class="headerlink" href="#StepwiseMutator.StepwiseMutator" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a stepwise mutation mutator that mutates an allele by
increasing or decreasing it. This mutator by default applies to
all loci unless parameter <em>loci</em> is specified. A single mutation
rate will be used for all loci if a single value of parameter
<em>rates</em> is given. Otherwise, a list of mutation rates can be
specified for each locus in parameter <em>loci</em>.</p>
<p>When a mutation event happens, this operator increases or
decreases an allele by <em>mutStep</em> steps. Acceptable input of
parameter <em>mutStep</em> include</p>
<ul class="simple">
<li>A number: This is the default mode with default value 1.</li>
<li><code class="docutils literal notranslate"><span class="pre">(GEOMETRIC_DISTRIBUTION,</span> <span class="pre">p)</span></code>: The number of steps follows a
a geometric distribution with parameter <em>p</em>.</li>
<li>A Python function: This user defined function accepts the
allele being mutated and return the steps to mutate.</li>
</ul>
<p>The mutation process is usually neutral in the sense that
mutating up and down is equally likely. You can adjust parameter
<em>incProb</em> to change this behavior.</p>
<p>If you need to use other generalized stepwise mutation models,
you can implement them using a <a class="reference internal" href="#PyMutator" title="PyMutator"><code class="xref py py-class docutils literal notranslate"><span class="pre">PyMutator</span></code></a>. If
performance becomes a concern, I may add them to this operator
if provided with a reliable reference.</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="class-pymutator">
<h2>class PyMutator<a class="headerlink" href="#class-pymutator" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="PyMutator">
<em class="property">class </em><code class="descname">PyMutator</code><a class="headerlink" href="#PyMutator" title="Permalink to this definition">¶</a></dt>
<dd><p>This hybrid mutator accepts a Python function that determines how
to mutate an allele when an mutation event happens.</p>
<dl class="method">
<dt id="PyMutator.PyMutator">
<code class="descname">PyMutator</code><span class="sig-paren">(</span><em>rates=[], loci=ALL_AVAIL, func=None, context=0, mapIn=[], mapOut=[], output=&quot;&quot;, begin=0, end=-1, step=1, at=[], reps=ALL_AVAIL, subPops=ALL_AVAIL, infoFields=[&quot;ind_id&quot;], lineageMode=FROM_INFO</em><span class="sig-paren">)</span><a class="headerlink" href="#PyMutator.PyMutator" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a hybrid mutator that uses a user-provided function to
mutate an allele when a mutation event happens. This function
(parameter <em>func</em>) accepts the allele to be mutated as parameter
<code class="docutils literal notranslate"><span class="pre">allele</span></code>, locus index <code class="docutils literal notranslate"><span class="pre">locus</span></code>, and optional array of alleles
as parameter <code class="docutils literal notranslate"><span class="pre">context</span></code>, which are <em>context</em> alleles the left
and right of the mutated allele. Invalid context alleles (e.g.
left allele to the first locus of a chromosome) will be marked
by -1. The return value of this function will be used to mutate
the passed allele. The passed, returned and context alleles
might be altered if parameter <em>mapIn</em> and <em>mapOut</em> are used.
This mutator by default applies to all loci unless parameter
<em>loci</em> is specified. A single mutation rate will be used for all
loci if a single value of parameter <em>rates</em> is given. Otherwise,
a list of mutation rates can be specified for each locus in
parameter <em>loci</em>. Please refer to classes <code class="docutils literal notranslate"><span class="pre">mutator</span></code> and
<a class="reference internal" href="refManual_ch3_sec1.html#BaseOperator" title="BaseOperator"><code class="xref py py-class docutils literal notranslate"><span class="pre">BaseOperator</span></code></a> for descriptions of other parameters.</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="class-mixedmutator">
<h2>class MixedMutator<a class="headerlink" href="#class-mixedmutator" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="MixedMutator">
<em class="property">class </em><code class="descname">MixedMutator</code><a class="headerlink" href="#MixedMutator" title="Permalink to this definition">¶</a></dt>
<dd><p>This mixed mutator accepts a list of mutators and use one of them
to mutate an allele when an mutation event happens.</p>
<dl class="method">
<dt id="MixedMutator.MixedMutator">
<code class="descname">MixedMutator</code><span class="sig-paren">(</span><em>rates=[], loci=ALL_AVAIL, mutators=[], prob=[], mapIn=[], mapOut=[], context=0, output=&quot;&quot;, begin=0, end=-1, step=1, at=[], reps=ALL_AVAIL, subPops=ALL_AVAIL, infoFields=[&quot;ind_id&quot;], lineageMode=FROM_INFO</em><span class="sig-paren">)</span><a class="headerlink" href="#MixedMutator.MixedMutator" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a mutator that randomly chooses one of the specified
<em>mutators</em> to mutate an allele when a mutation event happens.
The mutators are choosen according to a list of probabilities
(parameter <em>prob</em>) that should add up to <code class="docutils literal notranslate"><span class="pre">1</span></code>. The passed and
returned alleles might be changed if parameters <em>mapIn</em> and
<em>mapOut</em> are used. Most parameters, including <em>loci</em>, <em>mapIn</em>,
<em>mapOut</em>, <em>rep</em>, and <em>subPops</em> of mutators specified in
parameter <em>mutators</em> are ignored. This mutator by default
applies to all loci unless parameter <em>loci</em> is specified. Please
refer to classes <code class="docutils literal notranslate"><span class="pre">mutator</span></code> and <a class="reference internal" href="refManual_ch3_sec1.html#BaseOperator" title="BaseOperator"><code class="xref py py-class docutils literal notranslate"><span class="pre">BaseOperator</span></code></a> for
descriptions of other parameters.</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="class-contextmutator">
<h2>class ContextMutator<a class="headerlink" href="#class-contextmutator" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="ContextMutator">
<em class="property">class </em><code class="descname">ContextMutator</code><a class="headerlink" href="#ContextMutator" title="Permalink to this definition">¶</a></dt>
<dd><p>This context-dependent mutator accepts a list of mutators and use
one of them to mutate an allele depending on the context of the
mutated allele.</p>
<dl class="method">
<dt id="ContextMutator.ContextMutator">
<code class="descname">ContextMutator</code><span class="sig-paren">(</span><em>rates=[], loci=ALL_AVAIL, mutators=[], contexts=[], mapIn=[], mapOut=[], output=&quot;&quot;, begin=0, end=-1, step=1, at=[], reps=ALL_AVAIL, subPops=ALL_AVAIL, infoFields=[&quot;ind_id&quot;], lineageMode=FROM_INFO</em><span class="sig-paren">)</span><a class="headerlink" href="#ContextMutator.ContextMutator" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a mutator that choose one of the specified <em>mutators</em> to
mutate an allele when a mutation event happens. The mutators are
choosen according to the context of the mutated allele, which is
specified as a list of alleles to the left and right of an
allele (parameter <em>contexts</em>). For example, <code class="docutils literal notranslate"><span class="pre">contexts=[(0,0),</span>
<span class="pre">(0,1),</span> <span class="pre">(1,1)]</span></code> indicates which mutators should be used to
mutate allele <code class="docutils literal notranslate"><span class="pre">X</span></code> in the context of <code class="docutils literal notranslate"><span class="pre">0X0</span></code>, <code class="docutils literal notranslate"><span class="pre">0X1</span></code>, and
<code class="docutils literal notranslate"><span class="pre">1X1</span></code>. A context can include more than one alleles at both
left and right sides of a mutated allele but all contexts should
have the same (even) number of alleles. If an allele does not
have full context (e.g. when a locus is the first locus on a
chromosome), unavailable alleles will be marked as -1. There
should be a mutator for each context but an additional mutator
can be specified as the default mutator for unmatched contexts.
If parameters <em>mapIn</em> is specified, both mutated allele and its
context alleles will be mapped. Most parameters, including
<em>loci</em>, <em>mapIn</em>, <em>mapOut</em>, <em>rep</em>, and <em>subPops</em> of mutators
specified in parameter <em>mutators</em> are ignored. This mutator by
default applies to all loci unless parameter <em>loci</em> is
specified. Please refer to classes <code class="docutils literal notranslate"><span class="pre">mutator</span></code> and
<a class="reference internal" href="refManual_ch3_sec1.html#BaseOperator" title="BaseOperator"><code class="xref py py-class docutils literal notranslate"><span class="pre">BaseOperator</span></code></a> for descriptions of other parameters.</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="class-pointmutator">
<h2>class PointMutator<a class="headerlink" href="#class-pointmutator" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="PointMutator">
<em class="property">class </em><code class="descname">PointMutator</code><a class="headerlink" href="#PointMutator" title="Permalink to this definition">¶</a></dt>
<dd><p>A point mutator is different from all other mutators because
mutations in this mutator do not happen randomly. Instead, it
happens to specific loci and mutate an allele to a specific state,
regardless of its original state. This mutator is usually used to
introduce a mutant to a population.</p>
<dl class="method">
<dt id="PointMutator.PointMutator">
<code class="descname">PointMutator</code><span class="sig-paren">(</span><em>loci, allele, ploidy=0, inds=[], output=&quot;&quot;, begin=0, end=-1, step=1, at=[], reps=ALL_AVAIL, subPops=0, infoFields=[&quot;ind_id&quot;], lineageMode=FROM_INFO</em><span class="sig-paren">)</span><a class="headerlink" href="#PointMutator.PointMutator" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a point mutator that mutates alleles at specified <em>loci</em>
to a given <em>allele</em> of individuals <em>inds</em>. If there are multiple
alleles at a locus (e.g. individuals in a diploid population),
only the first allele is mutated unless indexes of alleles are
listed in parameter <em>ploidy</em>. This operator is by default
applied to individuals in the first subpopulation but you can
apply it to a different or more than one (virtual)
subpopulations using parameter <em>subPops</em> (<code class="docutils literal notranslate"><span class="pre">AllAvail</span></code> is also
accepted). Please refer to class <a class="reference internal" href="refManual_ch3_sec1.html#BaseOperator" title="BaseOperator"><code class="xref py py-class docutils literal notranslate"><span class="pre">BaseOperator</span></code></a> for
detailed descriptions of other parameters.</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="class-snpmutator">
<h2>class SNPMutator<a class="headerlink" href="#class-snpmutator" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="SNPMutator">
<em class="property">class </em><code class="descname">SNPMutator</code><a class="headerlink" href="#SNPMutator" title="Permalink to this definition">¶</a></dt>
<dd><p>A mutator model that assumes two alleles 0 and 1 and accepts mutation
rate from 0 to 1, and from 1 to 0 alleles.</p>
<dl class="method">
<dt id="SNPMutator.SNPMutator">
<code class="descname">SNPMutator</code><span class="sig-paren">(</span><em>u=0, v=0, loci=True, mapIn=[], mapOut=[], output='', begin=0, end=-1, step=1, at=[], reps=True, subPops=ALL_AVAIL, infoFields=['ind_id'], lineageMode=115</em><span class="sig-paren">)</span><a class="headerlink" href="#SNPMutator.SNPMutator" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a <a class="reference internal" href="#MatrixMutator" title="MatrixMutator"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixMutator</span></code></a> with proper mutate matrix for a two-allele
mutation model using mutation rate from allele 0 to 1 (parameter <code class="docutils literal notranslate"><span class="pre">u</span></code>)
and from 1 to 0 (parameter <code class="docutils literal notranslate"><span class="pre">v</span></code>)</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="class-acgtmutator">
<h2>class AcgtMutator<a class="headerlink" href="#class-acgtmutator" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="AcgtMutator">
<em class="property">class </em><code class="descname">AcgtMutator</code><a class="headerlink" href="#AcgtMutator" title="Permalink to this definition">¶</a></dt>
<dd><p>This mutation operator assumes alleles 0, 1, 2, 3 as nucleotides <code class="docutils literal notranslate"><span class="pre">A</span></code>,
<code class="docutils literal notranslate"><span class="pre">C</span></code>, <code class="docutils literal notranslate"><span class="pre">G</span></code> and <code class="docutils literal notranslate"><span class="pre">T</span></code> and use a 4 by 4 mutation rate matrix to mutate them.
Although a general model needs 12 parameters, less parameters are needed
for specific nucleotide mutation models (parameter <code class="docutils literal notranslate"><span class="pre">model</span></code>). The length
and meaning of parameter <code class="docutils literal notranslate"><span class="pre">rate</span></code> is model dependent.</p>
<dl class="method">
<dt id="AcgtMutator.AcgtMutator">
<code class="descname">AcgtMutator</code><span class="sig-paren">(</span><em>rate=[], model='general', loci=True, mapIn=[], mapOut=[], output='', begin=0, end=-1, step=1, at=[], reps=True, subPops=ALL_AVAIL, infoFields=['ind_id'], lineageMode=115</em><span class="sig-paren">)</span><a class="headerlink" href="#AcgtMutator.AcgtMutator" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a mutation model that mutates between nucleotides <code class="docutils literal notranslate"><span class="pre">A</span></code>,
<code class="docutils literal notranslate"><span class="pre">C</span></code>, <code class="docutils literal notranslate"><span class="pre">G</span></code>, and <code class="docutils literal notranslate"><span class="pre">T</span></code> (alleles are coded in that order as 0, 1, 2
and 3). Currently supported models are Jukes and Cantor 1969 model
(<code class="docutils literal notranslate"><span class="pre">JC69</span></code>), Kimura’s 2-parameter model (<code class="docutils literal notranslate"><span class="pre">K80</span></code>), Felsenstein 1981
model (<code class="docutils literal notranslate"><span class="pre">F81</span></code>), Hasgawa, Kishino and Yano 1985 model (<code class="docutils literal notranslate"><span class="pre">HKY85</span></code>),
Tamura 1992 model (<code class="docutils literal notranslate"><span class="pre">T92</span></code>), Tamura and Nei 1993 model (<code class="docutils literal notranslate"><span class="pre">TN93</span></code>),
Generalized time reversible model (<code class="docutils literal notranslate"><span class="pre">GTR</span></code>), and a general model
(<code class="docutils literal notranslate"><span class="pre">general</span></code>) with 12 parameters. Please refer to the simuPOP user’s
guide for detailed information about each model.</p>
</dd></dl>

</dd></dl>

</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">simuPOP</a></h1>








<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="userGuide_ch1.html">Front Matter</a></li>
<li class="toctree-l1"><a class="reference internal" href="userGuide_ch2.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="userGuide_ch3.html">Loading and running simuPOP</a></li>
<li class="toctree-l1"><a class="reference internal" href="userGuide_ch4.html">Individuals and Populations</a></li>
<li class="toctree-l1"><a class="reference internal" href="userGuide_ch5.html">simuPOP Operators</a></li>
<li class="toctree-l1"><a class="reference internal" href="userGuide_ch6.html">Evolving populations</a></li>
<li class="toctree-l1"><a class="reference internal" href="userGuide_ch7.html">Utility Modules</a></li>
<li class="toctree-l1"><a class="reference internal" href="userGuide_ch8.html">A real world example</a></li>
<li class="toctree-l1"><a class="reference internal" href="refManual_ch1.html">Front Matter</a></li>
<li class="toctree-l1"><a class="reference internal" href="refManual_ch2.html">simuPOP Components</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="refManual_ch3.html">Operator References</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="refManual_ch3_sec1.html">Base class for all operators</a></li>
<li class="toctree-l2"><a class="reference internal" href="refManual_ch3_sec2.html">Initialization</a></li>
<li class="toctree-l2"><a class="reference internal" href="refManual_ch3_sec3.html">Expression and Statements</a></li>
<li class="toctree-l2"><a class="reference internal" href="refManual_ch3_sec4.html">Demographic models</a></li>
<li class="toctree-l2"><a class="reference internal" href="refManual_ch3_sec5.html">Genotype transmitters</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Mutation</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#class-basemutator">class BaseMutator</a></li>
<li class="toctree-l3"><a class="reference internal" href="#class-matrixmutator">class MatrixMutator</a></li>
<li class="toctree-l3"><a class="reference internal" href="#class-kallelemutator">class KAlleleMutator</a></li>
<li class="toctree-l3"><a class="reference internal" href="#class-stepwisemutator">class StepwiseMutator</a></li>
<li class="toctree-l3"><a class="reference internal" href="#class-pymutator">class PyMutator</a></li>
<li class="toctree-l3"><a class="reference internal" href="#class-mixedmutator">class MixedMutator</a></li>
<li class="toctree-l3"><a class="reference internal" href="#class-contextmutator">class ContextMutator</a></li>
<li class="toctree-l3"><a class="reference internal" href="#class-pointmutator">class PointMutator</a></li>
<li class="toctree-l3"><a class="reference internal" href="#class-snpmutator">class SNPMutator</a></li>
<li class="toctree-l3"><a class="reference internal" href="#class-acgtmutator">class AcgtMutator</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="refManual_ch3_sec7.html">Penetrance</a></li>
<li class="toctree-l2"><a class="reference internal" href="refManual_ch3_sec8.html">Quantitative Trait</a></li>
<li class="toctree-l2"><a class="reference internal" href="refManual_ch3_sec9.html">Natural selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="refManual_ch3_sec10.html">Tagging operators</a></li>
<li class="toctree-l2"><a class="reference internal" href="refManual_ch3_sec11.html">Statistics Calculation</a></li>
<li class="toctree-l2"><a class="reference internal" href="refManual_ch3_sec12.html">Conditional operators</a></li>
<li class="toctree-l2"><a class="reference internal" href="refManual_ch3_sec13.html">The Python operator</a></li>
<li class="toctree-l2"><a class="reference internal" href="refManual_ch3_sec14.html">Miscellaneous operators</a></li>
<li class="toctree-l2"><a class="reference internal" href="refManual_ch3_sec15.html">Function form of operators</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="refManual_ch4.html">Utility Modules</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="refManual.html">&lt;no title&gt;</a><ul>
  <li><a href="refManual_ch3.html">Operator References</a><ul>
      <li>Previous: <a href="refManual_ch3_sec5.html" title="previous chapter">Genotype transmitters</a></li>
      <li>Next: <a href="refManual_ch3_sec7.html" title="next chapter">Penetrance</a></li>
  </ul></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2004-2019, Bo Peng.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.8.4</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/refManual_ch3_sec6.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>