
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Hybrid and Python operators &#8212; simuPOP  documentation</title>
    <link rel="stylesheet" href="_static/simuPOP.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Evolving populations" href="userGuide_ch6.html" />
    <link rel="prev" title="Miscellaneous operators" href="userGuide_ch5_sec13.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="hybrid-and-python-operators">
<h1>Hybrid and Python operators<a class="headerlink" href="#hybrid-and-python-operators" title="Permalink to this headline">¶</a></h1>
<div class="section" id="hybrid-operators">
<h2>Hybrid operators<a class="headerlink" href="#hybrid-operators" title="Permalink to this headline">¶</a></h2>
<p>Despite the large number of built-in operators, it is obviously not possible to
implement every genetics models available. For example, although simuPOP
provides several penetrance models, a user may want to try a customized one. In
this case, one can use a <em>hybrid operator</em>.</p>
<p>A <em>hybrid operator</em> is an operator that calls a user-defined function when its
applied to a population. The number and meaning of input parameters and return
values vary from operator to operator. For example, a hybrid mutator sends a to-
be-mutated allele to a user-defined function and use its return value as a
mutant allele. A hybrid selector uses the return value of a user defined
function as individual fitness. Such an operator handles the routine part of the
work (e.g. scan through a chromosome and determine which allele needs to be
mutated), and leave the creative part to users. Such a mutator can be used to
implement complicated genetic models such as an asymmetric stepwise mutation
model for microsatellite markers.</p>
<p><strong>simuPOP operators use parameter names to determine which information should be
passed to a user-defined function</strong>. For example, a hybrid quantitative trait
operator recognizes parameters <code class="docutils literal notranslate"><span class="pre">ind</span></code>, <code class="docutils literal notranslate"><span class="pre">geno</span></code>, <code class="docutils literal notranslate"><span class="pre">gen</span></code> and names of
information fields such as <code class="docutils literal notranslate"><span class="pre">smoking</span></code>. If your model depends on genotype, you
could provide a function with parameter geno (e.g. <code class="docutils literal notranslate"><span class="pre">func(geno)</span></code>); if your
model depends on smoking and genotype, you could provide a function with
parameters geno and smoking (e.g. <code class="docutils literal notranslate"><span class="pre">func(geno,</span> <span class="pre">smoking)</span></code>); if you model depends
on individual sex, you can use a function that passes the whole individual (e.g.
<code class="docutils literal notranslate"><span class="pre">func(ind)</span></code>) so that you could check individual sex. When a hybrid operator is
applied to a population, it will check the parameter names of provided Python
function and send requested information automatically.</p>
<p>For example, Example <a class="reference internal" href="#hybridoperator"><span class="std std-ref">hybridOperator</span></a> defines a three-
locus heterogeneity penetrance model Risch1990 that yields positive penetrance
only when at least two disease susceptibility alleles are available. The
underlying mechanism of this operator is that for each individual, simuPOP will
collect genotype at specified loci (parameter <code class="docutils literal notranslate"><span class="pre">loci</span></code>) and send them to
function <code class="docutils literal notranslate"><span class="pre">myPenetrance</span></code> and evaluate. The return values are used as the
penetrance value of the individual, which is then interpreted as the probability
that this individual will become affected.</p>
<p id="hybridoperator"><strong>Example</strong>: <em>Use a hybrid operator</em></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">simuPOP</span> <span class="k">as</span> <span class="nn">sim</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">def</span> <span class="nf">myPenetrance</span><span class="p">(</span><span class="n">geno</span><span class="p">):</span>
<span class="gp">... </span>    <span class="s1">&#39;A three-locus heterogeneity penetrance model&#39;</span>
<span class="gp">... </span>    <span class="k">if</span> <span class="nb">sum</span><span class="p">(</span><span class="n">geno</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">2</span><span class="p">:</span>
<span class="gp">... </span>        <span class="k">return</span> <span class="mi">0</span>
<span class="gp">... </span>    <span class="k">else</span><span class="p">:</span>
<span class="gp">... </span>        <span class="k">return</span> <span class="nb">sum</span><span class="p">(</span><span class="n">geno</span><span class="p">)</span><span class="o">*</span><span class="mf">0.1</span>
<span class="gp">...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">sim</span><span class="o">.</span><span class="n">Population</span><span class="p">(</span><span class="mi">1000</span><span class="p">,</span> <span class="n">loci</span><span class="o">=</span><span class="p">[</span><span class="mi">20</span><span class="p">]</span><span class="o">*</span><span class="mi">3</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span><span class="o">.</span><span class="n">evolve</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">initOps</span><span class="o">=</span><span class="p">[</span>
<span class="gp">... </span>        <span class="n">sim</span><span class="o">.</span><span class="n">InitSex</span><span class="p">(),</span>
<span class="gp">... </span>        <span class="n">sim</span><span class="o">.</span><span class="n">InitGenotype</span><span class="p">(</span><span class="n">freq</span><span class="o">=</span><span class="p">[</span><span class="mf">0.8</span><span class="p">,</span> <span class="mf">0.2</span><span class="p">])</span>
<span class="gp">... </span>    <span class="p">],</span>
<span class="gp">... </span>    <span class="n">matingScheme</span><span class="o">=</span><span class="n">sim</span><span class="o">.</span><span class="n">RandomMating</span><span class="p">(),</span>
<span class="gp">... </span>    <span class="n">postOps</span><span class="o">=</span><span class="p">[</span>
<span class="gp">... </span>        <span class="n">sim</span><span class="o">.</span><span class="n">PyPenetrance</span><span class="p">(</span><span class="n">func</span><span class="o">=</span><span class="n">myPenetrance</span><span class="p">,</span> <span class="n">loci</span><span class="o">=</span><span class="p">[</span><span class="mi">10</span><span class="p">,</span> <span class="mi">30</span><span class="p">,</span> <span class="mi">50</span><span class="p">]),</span>
<span class="gp">... </span>        <span class="n">sim</span><span class="o">.</span><span class="n">Stat</span><span class="p">(</span><span class="n">numOfAffected</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span>
<span class="gp">... </span>        <span class="n">sim</span><span class="o">.</span><span class="n">PyEval</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;&#39;</span><span class="si">%d</span><span class="s2">: </span><span class="si">%d</span><span class="s2">\n&#39; % (gen, numOfAffected)&quot;</span><span class="p">)</span>
<span class="gp">... </span>    <span class="p">],</span>
<span class="gp">... </span>    <span class="n">gen</span> <span class="o">=</span> <span class="mi">5</span>
<span class="gp">... </span><span class="p">)</span>
<span class="go">0: 97</span>
<span class="go">1: 96</span>
<span class="go">2: 78</span>
<span class="go">3: 95</span>
<span class="go">4: 80</span>
<span class="go">5</span>

<span class="go">now exiting runScriptInteractively...</span>
</pre></div>
</div>
<p><a class="reference external" href="hybrid.py">Download hybrid.py</a></p>
</div>
<div class="section" id="python-operator-pyoperator">
<h2>Python operator <a class="reference internal" href="refManual_ch3_sec13.html#PyOperator" title="PyOperator"><code class="xref py py-class docutils literal notranslate"><span class="pre">PyOperator</span></code></a> *<a class="headerlink" href="#python-operator-pyoperator" title="Permalink to this headline">¶</a></h2>
<p>If hybrid operators are still not flexible enough, you can always resort to a
pure-Python operator <a class="reference internal" href="refManual_ch3_sec13.html#PyOperator" title="PyOperator"><code class="xref py py-class docutils literal notranslate"><span class="pre">PyOperator</span></code></a>. This operator has full access to the
evolving population (or parents and offspring when aplied during-mating), and
can therefore perform arbitrary operations.</p>
<p>A <a class="reference internal" href="refManual_ch3_sec13.html#PyOperator" title="PyOperator"><code class="xref py py-class docutils literal notranslate"><span class="pre">PyOperator</span></code></a> that is applied pre- or post- mating expects a function
with one or both parameters <code class="docutils literal notranslate"><span class="pre">pop</span></code> and <code class="docutils literal notranslate"><span class="pre">param</span></code>, where<code class="docutils literal notranslate"><span class="pre">pop</span></code> is the
population being applied, and <code class="docutils literal notranslate"><span class="pre">param</span></code> is optional, depending on whether or not
a parameter is passed to the <a class="reference internal" href="refManual_ch3_sec13.html#PyOperator" title="PyOperator"><code class="xref py py-class docutils literal notranslate"><span class="pre">PyOperator</span></code></a>() constructor. Function
<code class="docutils literal notranslate"><span class="pre">func</span></code> can perform arbitrary action to <code class="docutils literal notranslate"><span class="pre">pop</span></code> and must return <code class="docutils literal notranslate"><span class="pre">True</span></code> or
<code class="docutils literal notranslate"><span class="pre">False</span></code>. <strong>The evolution of pop will be stopped if this function returns
False.</strong> This is essentially how operator <a class="reference internal" href="refManual_ch3_sec12.html#TerminateIf" title="TerminateIf"><code class="xref py py-class docutils literal notranslate"><span class="pre">TerminateIf</span></code></a> works.
Alternatively, this callback function can accept <code class="docutils literal notranslate"><span class="pre">ind</span></code> as one of the
parameters. In this case, the function will be called for all individuals or
individuals in specified (virtual) subpopulations. <strong>Individuals will be removed
from the populaton if this function returns False</strong>.</p>
<p>Example <a class="reference internal" href="#pyoperator"><span class="std std-ref">PyOperator</span></a> defines such a function. It accepts a
cutoff value and two mutation rates as parameters. It then calculate the
frequency of allele 1 at each locus and apply a two-allele model at high
mutation rate if the frequency is lower than the cutoff and a low mutation rate
otherwise. The <a class="reference internal" href="refManual_ch3_sec15.html#kAlleleMutate" title="kAlleleMutate"><code class="xref py py-func docutils literal notranslate"><span class="pre">kAlleleMutate</span></code></a> function is the function form of a mutator
<a class="reference internal" href="refManual_ch3_sec6.html#KAlleleMutator" title="KAlleleMutator"><code class="xref py py-class docutils literal notranslate"><span class="pre">KAlleleMutator</span></code></a> (see Section <a class="reference internal" href="userGuide_ch5_sec1.html#subsec-function-form"><span class="std std-ref">subsec_Function_form</span></a> for details).</p>
<p id="pyoperator"><strong>Example</strong>: <em>A frequency dependent mutation operator</em></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">simuPOP</span> <span class="k">as</span> <span class="nn">sim</span>
<span class="k">def</span> <span class="nf">dynaMutator</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="n">param</span><span class="p">):</span>
    <span class="sd">&#39;&#39;&#39;This mutator mutates commom loci with low mutation rate and rare</span>
<span class="sd">    loci with high mutation rate, as an attempt to raise allele frequency</span>
<span class="sd">    of rare loci to an higher level.&#39;&#39;&#39;</span>
    <span class="c1"># unpack parameter</span>
    <span class="p">(</span><span class="n">cutoff</span><span class="p">,</span> <span class="n">mu1</span><span class="p">,</span> <span class="n">mu2</span><span class="p">)</span> <span class="o">=</span> <span class="n">param</span><span class="p">;</span>
    <span class="n">sim</span><span class="o">.</span><span class="n">stat</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="n">alleleFreq</span><span class="o">=</span><span class="nb">range</span><span class="p">(</span><span class="n">pop</span><span class="o">.</span><span class="n">totNumLoci</span><span class="p">()))</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">pop</span><span class="o">.</span><span class="n">totNumLoci</span><span class="p">()):</span>
        <span class="c1"># Get the frequency of allele 1 (disease allele)</span>
        <span class="k">if</span> <span class="n">pop</span><span class="o">.</span><span class="n">dvars</span><span class="p">()</span><span class="o">.</span><span class="n">alleleFreq</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">cutoff</span><span class="p">:</span>
            <span class="n">sim</span><span class="o">.</span><span class="n">kAlleleMutate</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">rates</span><span class="o">=</span><span class="n">mu1</span><span class="p">,</span> <span class="n">loci</span><span class="o">=</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">sim</span><span class="o">.</span><span class="n">kAlleleMutate</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">rates</span><span class="o">=</span><span class="n">mu2</span><span class="p">,</span> <span class="n">loci</span><span class="o">=</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
    <span class="k">return</span> <span class="kc">True</span>
</pre></div>
</div>
<p><a class="reference external" href="PyOperator.py">Download PyOperator.py</a></p>
<p>Example <a class="reference internal" href="#usepyoperator"><span class="std std-ref">usePyOperator</span></a> demonstrates how to use this
operator. It first initializes the population using two <a class="reference internal" href="refManual_ch3_sec2.html#InitGenotype" title="InitGenotype"><code class="xref py py-class docutils literal notranslate"><span class="pre">InitGenotype</span></code></a>
operators that initialize loci with different allele frequencies. It applies a
<code class="docutils literal notranslate"><span class="pre">PyOperator</span></code>with function <code class="docutils literal notranslate"><span class="pre">dynaMutator</span></code> and a tuple of parameters. Allele
frequencies at all loci are printed at generation <code class="docutils literal notranslate"><span class="pre">0</span></code>, <code class="docutils literal notranslate"><span class="pre">10</span></code>, <code class="docutils literal notranslate"><span class="pre">20</span></code>, and
<code class="docutils literal notranslate"><span class="pre">30</span></code>. Note that this <a class="reference internal" href="refManual_ch3_sec13.html#PyOperator" title="PyOperator"><code class="xref py py-class docutils literal notranslate"><span class="pre">PyOperator</span></code></a> is applied at to the parental
generation so allele frequencies have to be recalculated to be used by post-
mating operator <a class="reference internal" href="refManual_ch3_sec3.html#PyEval" title="PyEval"><code class="xref py py-class docutils literal notranslate"><span class="pre">PyEval</span></code></a>.</p>
<p id="usepyoperator"><strong>Example</strong>: <em>Use a PyOperator during evolution</em></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">sim</span><span class="o">.</span><span class="n">Population</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="mi">10000</span><span class="p">,</span> <span class="n">loci</span><span class="o">=</span><span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span><span class="o">.</span><span class="n">evolve</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">initOps</span><span class="o">=</span><span class="p">[</span>
<span class="gp">... </span>        <span class="n">sim</span><span class="o">.</span><span class="n">InitSex</span><span class="p">(),</span>
<span class="gp">... </span>        <span class="n">sim</span><span class="o">.</span><span class="n">InitGenotype</span><span class="p">(</span><span class="n">freq</span><span class="o">=</span><span class="p">[</span><span class="o">.</span><span class="mi">99</span><span class="p">,</span> <span class="o">.</span><span class="mi">01</span><span class="p">],</span> <span class="n">loci</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">]),</span>
<span class="gp">... </span>        <span class="n">sim</span><span class="o">.</span><span class="n">InitGenotype</span><span class="p">(</span><span class="n">freq</span><span class="o">=</span><span class="p">[</span><span class="o">.</span><span class="mi">8</span><span class="p">,</span> <span class="o">.</span><span class="mi">2</span><span class="p">],</span> <span class="n">loci</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">])</span>
<span class="gp">... </span>    <span class="p">],</span>
<span class="gp">... </span>    <span class="n">preOps</span><span class="o">=</span><span class="n">sim</span><span class="o">.</span><span class="n">PyOperator</span><span class="p">(</span><span class="n">func</span><span class="o">=</span><span class="n">dynaMutator</span><span class="p">,</span> <span class="n">param</span><span class="o">=</span><span class="p">(</span><span class="o">.</span><span class="mi">2</span><span class="p">,</span> <span class="mf">1e-2</span><span class="p">,</span> <span class="mf">1e-5</span><span class="p">)),</span>
<span class="gp">... </span>    <span class="n">matingScheme</span><span class="o">=</span><span class="n">sim</span><span class="o">.</span><span class="n">RandomMating</span><span class="p">(),</span>
<span class="gp">... </span>    <span class="n">postOps</span><span class="o">=</span><span class="p">[</span>
<span class="gp">... </span>        <span class="n">sim</span><span class="o">.</span><span class="n">Stat</span><span class="p">(</span><span class="n">alleleFreq</span><span class="o">=</span><span class="nb">range</span><span class="p">(</span><span class="mi">5</span><span class="p">),</span> <span class="n">step</span><span class="o">=</span><span class="mi">10</span><span class="p">),</span>
<span class="gp">... </span>        <span class="n">sim</span><span class="o">.</span><span class="n">PyEval</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;&#39; &#39;.join([&#39;</span><span class="si">%.2f</span><span class="s2">&#39; % alleleFreq[x][1] for x in range(5)]) + &#39;\n&#39;&quot;</span><span class="p">,</span>
<span class="gp">... </span>            <span class="n">step</span><span class="o">=</span><span class="mi">10</span><span class="p">),</span>
<span class="gp">... </span>    <span class="p">],</span>
<span class="gp">... </span>    <span class="n">gen</span> <span class="o">=</span> <span class="mi">31</span>
<span class="gp">... </span><span class="p">)</span>
<span class="go">0.02 0.20 0.02 0.20 0.02</span>
<span class="go">0.11 0.22 0.11 0.20 0.11</span>
<span class="go">0.19 0.21 0.20 0.20 0.18</span>
<span class="go">0.21 0.21 0.22 0.21 0.21</span>
<span class="go">31</span>

<span class="go">now exiting runScriptInteractively...</span>
</pre></div>
</div>
<p><a class="reference external" href="PyOperator.py">Download PyOperator.py</a></p>
</div>
<div class="section" id="during-mating-python-operator">
<h2>During-mating Python operator *<a class="headerlink" href="#during-mating-python-operator" title="Permalink to this headline">¶</a></h2>
<p>A <a class="reference internal" href="refManual_ch3_sec13.html#PyOperator" title="PyOperator"><code class="xref py py-class docutils literal notranslate"><span class="pre">PyOperator</span></code></a> can also be applied during-mating. They can be used to
filter out unwanted offspring (by returning <code class="docutils literal notranslate"><span class="pre">False</span></code> in a user-defined
function), modify offspring, calculate statistics, or pass additional
information from parents to offspring. Depending the names of parameters of your
function, the Python operator will pass offspring (parameter <code class="docutils literal notranslate"><span class="pre">off</span></code>), his or
her parents (parameter <code class="docutils literal notranslate"><span class="pre">dad</span></code> and <code class="docutils literal notranslate"><span class="pre">mom</span></code>), the whole population (parameter
<code class="docutils literal notranslate"><span class="pre">pop</span></code>) and an optional parameter (parameter <code class="docutils literal notranslate"><span class="pre">param</span></code>) to this function. For
example, function <code class="docutils literal notranslate"><span class="pre">func(off)</span></code> will accept references to an offspring, and
<code class="docutils literal notranslate"><span class="pre">func(off,</span> <span class="pre">mom,</span> <span class="pre">dad)</span></code> will accept references to both offspring and his or her
parents.</p>
<p>Example <a class="reference internal" href="#duringmatingpyoperator"><span class="std std-ref">duringMatingPyOperator</span></a> demonstrates the
use of a during-mating Python operator. This operator rejects an offspring if it
has allele 1 at the first locus of the first homologous chromosome, and results
in an offspring population without such individuals.</p>
<p id="duringmatingpyoperator"><strong>Example</strong>: <em>Use a during-mating PyOperator</em></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">simuPOP</span> <span class="k">as</span> <span class="nn">sim</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">def</span> <span class="nf">rejectInd</span><span class="p">(</span><span class="n">off</span><span class="p">):</span>
<span class="gp">... </span>    <span class="s1">&#39;reject an individual if it off.allele(0) == 1&#39;</span>
<span class="gp">... </span>    <span class="k">return</span> <span class="n">off</span><span class="o">.</span><span class="n">allele</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span>
<span class="gp">...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">sim</span><span class="o">.</span><span class="n">Population</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">loci</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span><span class="o">.</span><span class="n">evolve</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">initOps</span><span class="o">=</span><span class="p">[</span>
<span class="gp">... </span>        <span class="n">sim</span><span class="o">.</span><span class="n">InitSex</span><span class="p">(),</span>
<span class="gp">... </span>        <span class="n">sim</span><span class="o">.</span><span class="n">InitGenotype</span><span class="p">(</span><span class="n">freq</span><span class="o">=</span><span class="p">[</span><span class="mf">0.5</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">])</span>
<span class="gp">... </span>    <span class="p">],</span>
<span class="gp">... </span>    <span class="n">matingScheme</span><span class="o">=</span><span class="n">sim</span><span class="o">.</span><span class="n">RandomMating</span><span class="p">(</span>
<span class="gp">... </span>        <span class="n">ops</span><span class="o">=</span><span class="p">[</span>
<span class="gp">... </span>            <span class="n">sim</span><span class="o">.</span><span class="n">MendelianGenoTransmitter</span><span class="p">(),</span>
<span class="gp">... </span>            <span class="n">sim</span><span class="o">.</span><span class="n">PyOperator</span><span class="p">(</span><span class="n">func</span><span class="o">=</span><span class="n">rejectInd</span><span class="p">)</span>
<span class="gp">... </span>        <span class="p">]),</span>
<span class="gp">... </span>    <span class="n">gen</span> <span class="o">=</span> <span class="mi">1</span>
<span class="gp">... </span><span class="p">)</span>
<span class="go">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># You should see no individual with allele 1 at locus 0, ploidy 0.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span><span class="o">.</span><span class="n">genotype</span><span class="p">()[:</span><span class="mi">20</span><span class="p">]</span>
<span class="go">[0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0]</span>

<span class="go">now exiting runScriptInteractively...</span>
</pre></div>
</div>
<p><a class="reference external" href="pyDuringMatingOperator.py">Download pyDuringMatingOperator.py</a></p>
<p><a class="reference internal" href="refManual_ch3_sec13.html#PyOperator" title="PyOperator"><code class="xref py py-class docutils literal notranslate"><span class="pre">PyOperator</span></code></a> is the most powerful operator in simuPOP and has been widely
used, for example, to calculate statistics and is not supported by the
<a class="reference internal" href="refManual_ch3_sec11.html#Stat" title="Stat"><code class="xref py py-class docutils literal notranslate"><span class="pre">Stat</span></code></a>() operator, to examine population property during evolution, or
prepare populations for a special mating scheme. However, because
<code class="docutils literal notranslate"><span class="pre">PyOperator</span></code>works in the Python interpreter, it is expected that it runs
slower than operators that are implemented at the C/C++ level. If performance
becomes an issue, you can re-implement part or all the operator in C++. Section
<a class="reference internal" href="userGuide_ch6_sec3.html#subsec-using-c"><span class="std std-ref">subsec_Using_C++</span></a> describes how to do this.</p>
</div>
<div class="section" id="define-your-own-operators">
<h2>Define your own operators *<a class="headerlink" href="#define-your-own-operators" title="Permalink to this headline">¶</a></h2>
<p><a class="reference internal" href="refManual_ch3_sec13.html#PyOperator" title="PyOperator"><code class="xref py py-class docutils literal notranslate"><span class="pre">PyOperator</span></code></a> is a Python class so you can derive your own operator from
this operator. The tricky part is that the constructor of the derived operator
needs to call the <code class="docutils literal notranslate"><span class="pre">__init__</span></code> function of <a class="reference internal" href="refManual_ch3_sec13.html#PyOperator" title="PyOperator"><code class="xref py py-class docutils literal notranslate"><span class="pre">PyOperator</span></code></a> will proper
functions. This technique has been used by simuPOP in a number of occasions. For
example, the <code class="docutils literal notranslate"><span class="pre">VarPlotter</span></code> operator defined in <code class="docutils literal notranslate"><span class="pre">plotter.py</span></code> is derived from
<a class="reference internal" href="refManual_ch3_sec13.html#PyOperator" title="PyOperator"><code class="xref py py-class docutils literal notranslate"><span class="pre">PyOperator</span></code></a>. This class encapsulates several different plot class that
uses <code class="docutils literal notranslate"><span class="pre">rpy</span></code> to plot python expressions. One of the plotters is passed to the
func parameter of <code class="docutils literal notranslate"><span class="pre">PyOperator.__init__</span></code> so that it can be called when this
operator is applied.</p>
<p>Example <a class="reference internal" href="userGuide_ch6_sec3.html#sequentialselfing"><span class="std std-ref">sequentialSelfing</span></a> rewrites the
<code class="docutils literal notranslate"><span class="pre">dynaMutator</span></code> defined in Example <a class="reference internal" href="#pyoperator"><span class="std std-ref">PyOperator</span></a> into a derived
operator. The parameters are now passed to the constructor of <code class="docutils literal notranslate"><span class="pre">dynaMutator</span></code>
and are saved as member variables. A member function <code class="docutils literal notranslate"><span class="pre">mutate</span></code> is defined and
is passed to the constructor of <a class="reference internal" href="refManual_ch3_sec13.html#PyOperator" title="PyOperator"><code class="xref py py-class docutils literal notranslate"><span class="pre">PyOperator</span></code></a>. Other than making
<code class="docutils literal notranslate"><span class="pre">dynaMutator</span></code> look like a real simuPOP operator, this example does not show a
lot of advantage over defining a function. However, when the operator gets
complicated (as in the case for <code class="docutils literal notranslate"><span class="pre">VarPlotter</span></code>), the object oriented
implementation will prevail.</p>
<p id="newoperator"><strong>Example</strong>: <em>Define a new Python operator</em></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">simuPOP</span> <span class="k">as</span> <span class="nn">sim</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">class</span> <span class="nc">dynaMutator</span><span class="p">(</span><span class="n">sim</span><span class="o">.</span><span class="n">PyOperator</span><span class="p">):</span>
<span class="gp">... </span>    <span class="sd">&#39;&#39;&#39;This mutator mutates commom loci with low mutation rate and rare</span>
<span class="gp">... </span><span class="sd">    loci with high mutation rate, as an attempt to raise allele frequency</span>
<span class="gp">... </span><span class="sd">    of rare loci to an higher level.&#39;&#39;&#39;</span>
<span class="gp">... </span>    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cutoff</span><span class="p">,</span> <span class="n">mu1</span><span class="p">,</span> <span class="n">mu2</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="gp">... </span>        <span class="bp">self</span><span class="o">.</span><span class="n">cutoff</span> <span class="o">=</span> <span class="n">cutoff</span>
<span class="gp">... </span>        <span class="bp">self</span><span class="o">.</span><span class="n">mu1</span> <span class="o">=</span> <span class="n">mu1</span>
<span class="gp">... </span>        <span class="bp">self</span><span class="o">.</span><span class="n">mu2</span> <span class="o">=</span> <span class="n">mu2</span>
<span class="gp">... </span>        <span class="n">sim</span><span class="o">.</span><span class="n">PyOperator</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">func</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">mutate</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
<span class="gp">... </span>    <span class="c1">#</span>
<span class="gp">... </span>    <span class="k">def</span> <span class="nf">mutate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">pop</span><span class="p">):</span>
<span class="gp">... </span>        <span class="n">sim</span><span class="o">.</span><span class="n">stat</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="n">alleleFreq</span><span class="o">=</span><span class="nb">range</span><span class="p">(</span><span class="n">pop</span><span class="o">.</span><span class="n">totNumLoci</span><span class="p">()))</span>
<span class="gp">... </span>        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">pop</span><span class="o">.</span><span class="n">totNumLoci</span><span class="p">()):</span>
<span class="gp">... </span>            <span class="c1"># Get the frequency of allele 1 (disease allele)</span>
<span class="gp">... </span>            <span class="k">if</span> <span class="n">pop</span><span class="o">.</span><span class="n">dvars</span><span class="p">()</span><span class="o">.</span><span class="n">alleleFreq</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">cutoff</span><span class="p">:</span>
<span class="gp">... </span>                <span class="n">sim</span><span class="o">.</span><span class="n">kAlleleMutate</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">rates</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">mu1</span><span class="p">,</span> <span class="n">loci</span><span class="o">=</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
<span class="gp">... </span>            <span class="k">else</span><span class="p">:</span>
<span class="gp">... </span>                <span class="n">sim</span><span class="o">.</span><span class="n">kAlleleMutate</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">rates</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">mu2</span><span class="p">,</span> <span class="n">loci</span><span class="o">=</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
<span class="gp">... </span>        <span class="k">return</span> <span class="kc">True</span>
<span class="gp">...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">sim</span><span class="o">.</span><span class="n">Population</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="mi">10000</span><span class="p">,</span> <span class="n">loci</span><span class="o">=</span><span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span><span class="o">.</span><span class="n">evolve</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">initOps</span><span class="o">=</span><span class="p">[</span>
<span class="gp">... </span>        <span class="n">sim</span><span class="o">.</span><span class="n">InitSex</span><span class="p">(),</span>
<span class="gp">... </span>        <span class="n">sim</span><span class="o">.</span><span class="n">InitGenotype</span><span class="p">(</span><span class="n">freq</span><span class="o">=</span><span class="p">[</span><span class="o">.</span><span class="mi">99</span><span class="p">,</span> <span class="o">.</span><span class="mi">01</span><span class="p">],</span> <span class="n">loci</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">]),</span>
<span class="gp">... </span>        <span class="n">sim</span><span class="o">.</span><span class="n">InitGenotype</span><span class="p">(</span><span class="n">freq</span><span class="o">=</span><span class="p">[</span><span class="o">.</span><span class="mi">8</span><span class="p">,</span> <span class="o">.</span><span class="mi">2</span><span class="p">],</span> <span class="n">loci</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">])</span>
<span class="gp">... </span>    <span class="p">],</span>
<span class="gp">... </span>    <span class="n">preOps</span><span class="o">=</span><span class="n">dynaMutator</span><span class="p">(</span><span class="n">cutoff</span><span class="o">=.</span><span class="mi">2</span><span class="p">,</span> <span class="n">mu1</span><span class="o">=</span><span class="mf">1e-2</span><span class="p">,</span> <span class="n">mu2</span><span class="o">=</span><span class="mf">1e-5</span><span class="p">),</span>
<span class="gp">... </span>    <span class="n">matingScheme</span><span class="o">=</span><span class="n">sim</span><span class="o">.</span><span class="n">RandomMating</span><span class="p">(),</span>
<span class="gp">... </span>    <span class="n">postOps</span><span class="o">=</span><span class="p">[</span>
<span class="gp">... </span>        <span class="n">sim</span><span class="o">.</span><span class="n">Stat</span><span class="p">(</span><span class="n">alleleFreq</span><span class="o">=</span><span class="nb">range</span><span class="p">(</span><span class="mi">5</span><span class="p">),</span> <span class="n">step</span><span class="o">=</span><span class="mi">10</span><span class="p">),</span>
<span class="gp">... </span>        <span class="n">sim</span><span class="o">.</span><span class="n">PyEval</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;&#39; &#39;.join([&#39;</span><span class="si">%.2f</span><span class="s2">&#39; % alleleFreq[x][1] for x in range(5)]) + &#39;\n&#39;&quot;</span><span class="p">,</span>
<span class="gp">... </span>            <span class="n">step</span><span class="o">=</span><span class="mi">10</span><span class="p">),</span>
<span class="gp">... </span>    <span class="p">],</span>
<span class="gp">... </span>    <span class="n">gen</span> <span class="o">=</span> <span class="mi">31</span>
<span class="gp">... </span><span class="p">)</span>
<span class="go">0.02 0.20 0.02 0.20 0.02</span>
<span class="go">0.11 0.22 0.11 0.20 0.11</span>
<span class="go">0.19 0.21 0.20 0.20 0.18</span>
<span class="go">0.21 0.21 0.22 0.21 0.21</span>
<span class="go">31</span>

<span class="go">now exiting runScriptInteractively...</span>
</pre></div>
</div>
<p><a class="reference external" href="newOperator.py">Download newOperator.py</a></p>
<p>New during-mating operators can be defined similarly. They are usually used to
define customized genotype transmitters. Section
<span class="xref std std-ref">subsec_Customized_genotype_transmitter</span> will describe this feature in detail.</p>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">simuPOP</a></h1>








<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="userGuide_ch1.html">Front Matter</a></li>
<li class="toctree-l1"><a class="reference internal" href="userGuide_ch2.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="userGuide_ch3.html">Loading and running simuPOP</a></li>
<li class="toctree-l1"><a class="reference internal" href="userGuide_ch4.html">Individuals and Populations</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="userGuide_ch5.html">simuPOP Operators</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch5_sec1.html">Introduction to operators</a></li>
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch5_sec2.html">Initialization</a></li>
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch5_sec3.html">Expressions and statements</a></li>
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch5_sec4.html">Demographic changes</a></li>
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch5_sec5.html">Genotype transmitters</a></li>
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch5_sec6.html">Mutation</a></li>
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch5_sec7.html">Penetrance</a></li>
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch5_sec8.html">Quantitative trait</a></li>
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch5_sec9.html">Natural Selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch5_sec10.html">Tagging operators</a></li>
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch5_sec11.html">Statistics calculation (operator <code class="docutils literal notranslate"><span class="pre">Stat</span></code>)</a></li>
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch5_sec12.html">Conditional operators</a></li>
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch5_sec13.html">Miscellaneous operators</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Hybrid and Python operators</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#hybrid-operators">Hybrid operators</a></li>
<li class="toctree-l3"><a class="reference internal" href="#python-operator-pyoperator">Python operator <code class="docutils literal notranslate"><span class="pre">PyOperator</span></code> *</a></li>
<li class="toctree-l3"><a class="reference internal" href="#during-mating-python-operator">During-mating Python operator *</a></li>
<li class="toctree-l3"><a class="reference internal" href="#define-your-own-operators">Define your own operators *</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="userGuide_ch6.html">Evolving populations</a></li>
<li class="toctree-l1"><a class="reference internal" href="userGuide_ch7.html">Utility Modules</a></li>
<li class="toctree-l1"><a class="reference internal" href="userGuide_ch8.html">A real world example</a></li>
<li class="toctree-l1"><a class="reference internal" href="refManual_ch1.html">Front Matter</a></li>
<li class="toctree-l1"><a class="reference internal" href="refManual_ch2.html">simuPOP Components</a></li>
<li class="toctree-l1"><a class="reference internal" href="refManual_ch3.html">Operator References</a></li>
<li class="toctree-l1"><a class="reference internal" href="refManual_ch4.html">Utility Modules</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="userGuide.html">&lt;no title&gt;</a><ul>
  <li><a href="userGuide_ch5.html">simuPOP Operators</a><ul>
      <li>Previous: <a href="userGuide_ch5_sec13.html" title="previous chapter">Miscellaneous operators</a></li>
      <li>Next: <a href="userGuide_ch6.html" title="next chapter">Evolving populations</a></li>
  </ul></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2004-2019, Bo Peng.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.8.4</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/userGuide_ch5_sec14.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>