
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Tracing allelic lineage * &#8212; simuPOP  documentation</title>
    <link rel="stylesheet" href="_static/simuPOP.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Pedigrees" href="userGuide_ch6_sec6.html" />
    <link rel="prev" title="Age structured populations with overlapping generations **" href="userGuide_ch6_sec4.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="tracing-allelic-lineage">
<h1>Tracing allelic lineage *<a class="headerlink" href="#tracing-allelic-lineage" title="Permalink to this headline">¶</a></h1>
<p>Lineage of alleles consists of information such as the distribution of alleles
(how many people carry this allele, and the relationship between carriers) and
age of alleles (when the alleles were introduced to the population). These
information are important for the study of evolutionary history of mutants. They
are not readily available for normal simulations, and even if you can track the
generations when mutants are introduced, alleles in the present generation that
are of the same type (Identity by Stat, IBS) do not necessarily have the same
ancestral origin (Identity by Decent, IBD).</p>
<p>The lineage modules of simuPOP provides facilities to track allelic lineage.
More specifically,</p>
<ul class="simple">
<li>Each allele is associated with an integer number (an allelic lineage) that
identifies the origin, or the source of the allele.</li>
<li>The lineage of each allele is transmitted along with the allele during
evolution. New alleles will be introduced with their own lineage, even if they
share the same states with existing alleles.</li>
<li>Origin of alleles can be accessed using member functions of the
<a class="reference internal" href="refManual_ch2_sec1.html#Individual" title="Individual"><code class="xref py py-class docutils literal notranslate"><span class="pre">Individual</span></code></a> and <a class="reference internal" href="refManual_ch2_sec1.html#Population" title="Population"><code class="xref py py-class docutils literal notranslate"><span class="pre">Population</span></code></a> classes.</li>
</ul>
<p>Example <a class="reference internal" href="#geneticcontribution"><span class="std std-ref">geneticContribution</span></a> demonstrates how to
determine the contribution of genetic information from each ancestor. For this
simulation, the alleles of each ancestor are associated with individual-specific
numbers. During evolution, some alleles might get lost, some are copied, and
pieces of chromosomes are mixed due to genetic recombination. At the end of
simulation, the average number of ‘contributors’ of genetic information to each
individual is calculated, as well as the percent of genetic information from
each ancestor. Although this particular simulation can be mimicked using pure-
genotype simulations by using special alleles for each ancestor, the combined
information regarding the state and origin of each allele will be very useful
for genetic studies that involve IBD and IBS.</p>
<p id="geneticcontribution"><strong>Example</strong>: <em>Contribution of genetic information from ancestors</em></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">simuOpt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">simuOpt</span><span class="o">.</span><span class="n">setOptions</span><span class="p">(</span><span class="n">alleleType</span><span class="o">=</span><span class="s1">&#39;lineage&#39;</span><span class="p">,</span> <span class="n">quiet</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">simuPOP</span> <span class="k">as</span> <span class="nn">sim</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">sim</span><span class="o">.</span><span class="n">Population</span><span class="p">(</span><span class="mi">1000</span><span class="p">,</span> <span class="n">loci</span><span class="o">=</span><span class="p">[</span><span class="mi">10</span><span class="p">]</span><span class="o">*</span><span class="mi">4</span><span class="p">)</span>
<span class="go">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span><span class="o">.</span><span class="n">evolve</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">initOps</span><span class="o">=</span><span class="p">[</span>
<span class="gp">... </span>        <span class="n">sim</span><span class="o">.</span><span class="n">InitSex</span><span class="p">(),</span>
<span class="gp">... </span>        <span class="n">sim</span><span class="o">.</span><span class="n">InitGenotype</span><span class="p">(</span><span class="n">freq</span><span class="o">=</span><span class="p">[</span><span class="mf">0.25</span><span class="p">]</span><span class="o">*</span><span class="mi">4</span><span class="p">),</span>
<span class="gp">... </span>        <span class="n">sim</span><span class="o">.</span><span class="n">InitLineage</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1000</span><span class="p">),</span> <span class="n">mode</span><span class="o">=</span><span class="n">sim</span><span class="o">.</span><span class="n">PER_INDIVIDUAL</span><span class="p">),</span>
<span class="gp">... </span>    <span class="p">],</span>
<span class="gp">... </span>    <span class="n">matingScheme</span><span class="o">=</span><span class="n">sim</span><span class="o">.</span><span class="n">RandomMating</span><span class="p">(</span><span class="n">ops</span><span class="o">=</span><span class="n">sim</span><span class="o">.</span><span class="n">Recombinator</span><span class="p">(</span><span class="n">rates</span><span class="o">=</span><span class="mf">0.001</span><span class="p">)),</span>
<span class="gp">... </span>    <span class="n">gen</span> <span class="o">=</span> <span class="mi">100</span>
<span class="gp">... </span><span class="p">)</span>
<span class="go">100</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># average number of &#39;contributors&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">num_contributors</span> <span class="o">=</span> <span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">ind</span><span class="o">.</span><span class="n">lineage</span><span class="p">()))</span> <span class="k">for</span> <span class="n">ind</span> <span class="ow">in</span> <span class="n">pop</span><span class="o">.</span><span class="n">individuals</span><span class="p">()]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Average number of contributors is </span><span class="si">%.2f</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="nb">sum</span><span class="p">(</span><span class="n">num_contributors</span><span class="p">)</span> <span class="o">/</span> <span class="nb">float</span><span class="p">(</span><span class="n">pop</span><span class="o">.</span><span class="n">popSize</span><span class="p">())))</span>
<span class="go">Average number of contributors is 13.98</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># percent of genetic information from each ancestor (baseline is 1/1000)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lineage</span> <span class="o">=</span> <span class="n">pop</span><span class="o">.</span><span class="n">lineage</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lin_perc</span> <span class="o">=</span> <span class="p">[</span><span class="n">lineage</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="o">/</span><span class="nb">float</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">lineage</span><span class="p">))</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1000</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># how many of ancestors do not have any allele left?</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Number of ancestors with no allele left: </span><span class="si">%d</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">lin_perc</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="mf">0.</span><span class="p">))</span>
<span class="go">Number of ancestors with no allele left: 817</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># top five contributors</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lin_perc</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lin_perc</span><span class="o">.</span><span class="n">reverse</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Top contributors (started with 0.001): </span><span class="si">%.5f</span><span class="s1"> </span><span class="si">%.5f</span><span class="s1"> </span><span class="si">%.5f</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">lin_perc</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">lin_perc</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">lin_perc</span><span class="p">[</span><span class="mi">2</span><span class="p">]))</span>
<span class="go">Top contributors (started with 0.001): 0.03474 0.03058 0.02475</span>

<span class="go">now exiting runScriptInteractively...</span>
</pre></div>
</div>
<p><a class="reference external" href="geneticContribution.py">Download geneticContribution.py</a></p>
<p>Example <a class="reference internal" href="#geneticcontribution"><span class="std std-ref">geneticContribution</span></a> uses operator
<a class="reference internal" href="refManual_ch3_sec2.html#InitLineage" title="InitLineage"><code class="xref py py-class docutils literal notranslate"><span class="pre">InitLineage</span></code></a> to explictly assign lineage to alleles of each individual.
You can also track the fate of finer genetic pieces by assigning different
lineage values to chromosomes, or each loci using different <code class="docutils literal notranslate"><span class="pre">mode</span></code>. This
operator can also assign lineage of alleles to an ID stored in an information
field, which is usually <code class="docutils literal notranslate"><span class="pre">ind_id</span></code>, a field used by operators such as
<a class="reference internal" href="refManual_ch3_sec10.html#IdTagger" title="IdTagger"><code class="xref py py-class docutils literal notranslate"><span class="pre">IdTagger</span></code></a> and <a class="reference internal" href="refManual_ch3_sec10.html#PedigreeTagger" title="PedigreeTagger"><code class="xref py py-class docutils literal notranslate"><span class="pre">PedigreeTagger</span></code></a> to assign and trace the pedigree
(parentship) information during evolution. More interesting, when such a field
is present, mutation operators will assign the IDs of recipients of mutants as
the lineage of these mutants. This makes it possible to track the origin of
mutants. Moreover, when a mode <code class="docutils literal notranslate"><span class="pre">FROM_INFO_SIGNED</span></code> is used, additional ploidy
information will be tagged to lineage values (negative values for mutants on the
second homologous copy of chromosomes) so that you can track the inheritance of
haplotypes.</p>
<p>To make use of these features, it is important to assign IDs to individuals
before these operators are applied. Example <a class="reference internal" href="#ageofmutants"><span class="std std-ref">ageOfMutants</span></a>
demonstrates how to use the lineage information to determine the age of mutants.
This example evolves a constant population of size 10,000. An <a class="reference internal" href="refManual_ch3_sec10.html#IdTagger" title="IdTagger"><code class="xref py py-class docutils literal notranslate"><span class="pre">IdTagger</span></code></a>
is used before <a class="reference internal" href="refManual_ch3_sec2.html#InitGenotype" title="InitGenotype"><code class="xref py py-class docutils literal notranslate"><span class="pre">InitGenotype</span></code></a> so individual IDs will be assigned as
allelic lineages. Because all offspring get their own IDs during evolution, the
IDs of individuals are assigned to mutants as their lineages, and can be used to
determine the age of these mutants. This is pretty easy to do in this example
because of constant population size. For more complex demographic models, you
might have to record the minimal and maximum IDs of each generation in order to
determine the age of mutants.</p>
<p id="ageofmutants"><strong>Example</strong>: <em>Distribution of age of mutants</em></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">simuOpt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">simuOpt</span><span class="o">.</span><span class="n">setOptions</span><span class="p">(</span><span class="n">alleleType</span><span class="o">=</span><span class="s1">&#39;lineage&#39;</span><span class="p">,</span> <span class="n">quiet</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">simuPOP</span> <span class="k">as</span> <span class="nn">sim</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">sim</span><span class="o">.</span><span class="n">Population</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="mi">10000</span><span class="p">,</span> <span class="n">loci</span><span class="o">=</span><span class="p">[</span><span class="mi">10</span><span class="p">]</span><span class="o">*</span><span class="mi">10</span><span class="p">,</span> <span class="n">infoFields</span><span class="o">=</span><span class="s1">&#39;ind_id&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># just to make sure IDs starts from 1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sim</span><span class="o">.</span><span class="n">IdTagger</span><span class="p">()</span><span class="o">.</span><span class="n">reset</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span><span class="o">.</span><span class="n">evolve</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">initOps</span> <span class="o">=</span> <span class="p">[</span>
<span class="gp">... </span>        <span class="n">sim</span><span class="o">.</span><span class="n">InitSex</span><span class="p">(),</span>
<span class="gp">... </span>        <span class="n">sim</span><span class="o">.</span><span class="n">InitGenotype</span><span class="p">(</span><span class="n">freq</span><span class="o">=</span><span class="p">[</span><span class="mf">0.2</span><span class="p">,</span> <span class="mf">0.3</span><span class="p">,</span> <span class="mf">0.4</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">]),</span>
<span class="gp">... </span>        <span class="n">sim</span><span class="o">.</span><span class="n">IdTagger</span><span class="p">(),</span>
<span class="gp">... </span>        <span class="n">sim</span><span class="o">.</span><span class="n">InitLineage</span><span class="p">(</span><span class="n">mode</span><span class="o">=</span><span class="n">sim</span><span class="o">.</span><span class="n">FROM_INFO</span><span class="p">),</span>
<span class="gp">... </span>    <span class="p">],</span>
<span class="gp">... </span>    <span class="c1"># an extremely high mutation rate, just for demonstration</span>
<span class="gp">... </span>    <span class="n">preOps</span> <span class="o">=</span> <span class="n">sim</span><span class="o">.</span><span class="n">AcgtMutator</span><span class="p">(</span><span class="n">rate</span><span class="o">=</span><span class="mf">0.01</span><span class="p">,</span> <span class="n">model</span><span class="o">=</span><span class="s1">&#39;JC69&#39;</span><span class="p">),</span>
<span class="gp">... </span>    <span class="n">matingScheme</span><span class="o">=</span><span class="n">sim</span><span class="o">.</span><span class="n">RandomMating</span><span class="p">(</span>
<span class="gp">... </span>        <span class="n">ops</span><span class="o">=</span><span class="p">[</span>
<span class="gp">... </span>            <span class="n">sim</span><span class="o">.</span><span class="n">IdTagger</span><span class="p">(),</span>
<span class="gp">... </span>            <span class="n">sim</span><span class="o">.</span><span class="n">MendelianGenoTransmitter</span><span class="p">(),</span>
<span class="gp">... </span>        <span class="p">]</span>
<span class="gp">... </span>    <span class="p">),</span>
<span class="gp">... </span>    <span class="n">gen</span> <span class="o">=</span> <span class="mi">10</span>
<span class="gp">... </span><span class="p">)</span>
<span class="go">10</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lin</span> <span class="o">=</span> <span class="n">pop</span><span class="o">.</span><span class="n">lineage</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Number of alleles from each generation</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">gen</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">10</span><span class="p">):</span>
<span class="gp">... </span>    <span class="n">id_start</span> <span class="o">=</span> <span class="n">gen</span><span class="o">*</span><span class="mi">10000</span> <span class="o">+</span> <span class="mi">1</span>
<span class="gp">... </span>    <span class="n">id_end</span> <span class="o">=</span> <span class="p">(</span><span class="n">gen</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span><span class="o">*</span><span class="mi">10000</span>
<span class="gp">... </span>    <span class="n">num_mut</span> <span class="o">=</span> <span class="nb">len</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">lin</span> <span class="k">if</span> <span class="n">x</span> <span class="o">&gt;=</span> <span class="n">id_start</span> <span class="ow">and</span> <span class="n">x</span> <span class="o">&lt;=</span> <span class="n">id_end</span><span class="p">])</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Gen </span><span class="si">%d</span><span class="s1">: </span><span class="si">%5.2f</span><span class="s1"> </span><span class="si">%%</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">gen</span><span class="p">,</span> <span class="n">num_mut</span> <span class="o">/</span> <span class="p">(</span><span class="mi">2</span><span class="o">*</span><span class="mi">10000</span><span class="o">*</span><span class="mf">100.</span><span class="p">)</span> <span class="o">*</span> <span class="mi">100</span><span class="p">))</span>
<span class="gp">...</span>
<span class="go">Gen 0: 93.40 %</span>
<span class="go">Gen 1:  0.72 %</span>
<span class="go">Gen 2:  0.71 %</span>
<span class="go">Gen 3:  0.70 %</span>
<span class="go">Gen 4:  0.74 %</span>
<span class="go">Gen 5:  0.76 %</span>
<span class="go">Gen 6:  0.73 %</span>
<span class="go">Gen 7:  0.74 %</span>
<span class="go">Gen 8:  0.75 %</span>
<span class="go">Gen 9:  0.75 %</span>

<span class="go">now exiting runScriptInteractively...</span>
</pre></div>
</div>
<p><a class="reference external" href="ageOfMutants.py">Download ageOfMutants.py</a></p>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">simuPOP</a></h1>








<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="userGuide_ch1.html">Front Matter</a></li>
<li class="toctree-l1"><a class="reference internal" href="userGuide_ch2.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="userGuide_ch3.html">Loading and running simuPOP</a></li>
<li class="toctree-l1"><a class="reference internal" href="userGuide_ch4.html">Individuals and Populations</a></li>
<li class="toctree-l1"><a class="reference internal" href="userGuide_ch5.html">simuPOP Operators</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="userGuide_ch6.html">Evolving populations</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch6_sec1.html">Mating Schemes</a></li>
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch6_sec2.html">Simulator</a></li>
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch6_sec3.html">Non-random and customized mating schemes *</a></li>
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch6_sec4.html">Age structured populations with overlapping generations **</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Tracing allelic lineage *</a></li>
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch6_sec6.html">Pedigrees</a></li>
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch6_sec7.html">Evolve a population following a specified pedigree structure **</a></li>
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch6_sec8.html">Simulation of mitochondrial DNAs (mtDNAs) *</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="userGuide_ch7.html">Utility Modules</a></li>
<li class="toctree-l1"><a class="reference internal" href="userGuide_ch8.html">A real world example</a></li>
<li class="toctree-l1"><a class="reference internal" href="refManual_ch1.html">Front Matter</a></li>
<li class="toctree-l1"><a class="reference internal" href="refManual_ch2.html">simuPOP Components</a></li>
<li class="toctree-l1"><a class="reference internal" href="refManual_ch3.html">Operator References</a></li>
<li class="toctree-l1"><a class="reference internal" href="refManual_ch4.html">Utility Modules</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="userGuide.html">&lt;no title&gt;</a><ul>
  <li><a href="userGuide_ch6.html">Evolving populations</a><ul>
      <li>Previous: <a href="userGuide_ch6_sec4.html" title="previous chapter">Age structured populations with overlapping generations **</a></li>
      <li>Next: <a href="userGuide_ch6_sec6.html" title="next chapter">Pedigrees</a></li>
  </ul></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2004-2019, Bo Peng.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.8.4</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/userGuide_ch6_sec5.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>