
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Initialization &#8212; simuPOP  documentation</title>
    <link rel="stylesheet" href="_static/simuPOP.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Expressions and statements" href="userGuide_ch5_sec3.html" />
    <link rel="prev" title="Introduction to operators" href="userGuide_ch5_sec1.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="initialization">
<h1>Initialization<a class="headerlink" href="#initialization" title="Permalink to this headline">¶</a></h1>
<p>simuPOP provides three operators to initialize individual sex, information
fields and genotype at the population level. A number of parameter are provided
to cover most commonly used initialization scenarios. A Python operator can be
used to intialize a population explicitly if none of the operators fits your
need.</p>
<div class="section" id="initialize-individual-sex-operator-initsex">
<h2>Initialize individual sex (operator <a class="reference internal" href="refManual_ch3_sec2.html#InitSex" title="InitSex"><code class="xref py py-class docutils literal notranslate"><span class="pre">InitSex</span></code></a>)<a class="headerlink" href="#initialize-individual-sex-operator-initsex" title="Permalink to this headline">¶</a></h2>
<p>Operator <a class="reference internal" href="refManual_ch3_sec2.html#InitSex" title="InitSex"><code class="xref py py-class docutils literal notranslate"><span class="pre">InitSex</span></code></a>() and function <code class="docutils literal notranslate"><span class="pre">initSex()</span></code> initialize individual
sex either randomly or using a given sequence. In the first case, individuals
are assigned <code class="docutils literal notranslate"><span class="pre">MALE</span></code> or <code class="docutils literal notranslate"><span class="pre">FEMALE</span></code> with equal probability unless parameter
<em>maleFreq</em> is used to specify the probability of having a male Individual.
Alternatively, parameter <em>maleProp</em> can be used to specify exact proportions of
male individuals so that you will have exactly 1000 males and 1000 females if
you apply <a class="reference internal" href="refManual_ch3_sec2.html#InitSex" title="InitSex"><code class="xref py py-class docutils literal notranslate"><span class="pre">InitSex</span></code></a>(<code class="docutils literal notranslate"><span class="pre">maleProp=0.5</span></code>) to a population of 2000
individuals.</p>
<p>Both parameters <code class="docutils literal notranslate"><span class="pre">maleFreq</span></code> and <code class="docutils literal notranslate"><span class="pre">maleProp</span></code> assigns individual sex randomly.
If for some reason you need to specify individual sex explicitly, you could use
a sequence of sex (<code class="docutils literal notranslate"><span class="pre">MALE</span></code> or <code class="docutils literal notranslate"><span class="pre">FEMALE</span></code>) to assign sex to individuals
succesively. The list will be reused if needed. If a list of (virtual)
subpopulations are given, this operator will only initialize individuals in
these (virtual) subpopulations. Example <a class="reference internal" href="#initsex"><span class="std std-ref">InitSex</span></a> demonstrates
how to use two <a class="reference internal" href="refManual_ch3_sec2.html#InitSex" title="InitSex"><code class="xref py py-class docutils literal notranslate"><span class="pre">InitSex</span></code></a> operators to initialize two subpopulations.</p>
<p id="initsex"><strong>Example</strong>: <em>Initialize individual sex</em></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">simuPOP</span> <span class="k">as</span> <span class="nn">sim</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">sim</span><span class="o">.</span><span class="n">Population</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="p">[</span><span class="mi">1000</span><span class="p">,</span> <span class="mi">1000</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sim</span><span class="o">.</span><span class="n">initSex</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="n">maleFreq</span><span class="o">=</span><span class="mf">0.3</span><span class="p">,</span> <span class="n">subPops</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sim</span><span class="o">.</span><span class="n">initSex</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="n">sex</span><span class="o">=</span><span class="p">[</span><span class="n">sim</span><span class="o">.</span><span class="n">MALE</span><span class="p">,</span> <span class="n">sim</span><span class="o">.</span><span class="n">FEMALE</span><span class="p">,</span> <span class="n">sim</span><span class="o">.</span><span class="n">FEMALE</span><span class="p">],</span> <span class="n">subPops</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sim</span><span class="o">.</span><span class="n">stat</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="n">numOfMales</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="nb">vars</span><span class="o">=</span><span class="s1">&#39;numOfMales_sp&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">pop</span><span class="o">.</span><span class="n">dvars</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">numOfMales</span><span class="p">)</span>
<span class="go">290</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">pop</span><span class="o">.</span><span class="n">dvars</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">numOfMales</span><span class="p">)</span>
<span class="go">334</span>

<span class="go">now exiting runScriptInteractively...</span>
</pre></div>
</div>
<p><a class="reference external" href="InitSex.py">Download InitSex.py</a></p>
</div>
<div class="section" id="initialize-genotype-operator-initgenotype">
<h2>Initialize genotype (operator <a class="reference internal" href="refManual_ch3_sec2.html#InitGenotype" title="InitGenotype"><code class="xref py py-class docutils literal notranslate"><span class="pre">InitGenotype</span></code></a>)<a class="headerlink" href="#initialize-genotype-operator-initgenotype" title="Permalink to this headline">¶</a></h2>
<p>Operator <a class="reference internal" href="refManual_ch3_sec2.html#InitGenotype" title="InitGenotype"><code class="xref py py-class docutils literal notranslate"><span class="pre">InitGenotype</span></code></a> (and its function form <code class="docutils literal notranslate"><span class="pre">initGenotype</span></code>)
initializes individual genotype by allele frequency, allele proportion,
haplotype frequency, haplotype proportions or a list of genotypes:</p>
<ul class="simple">
<li>By frequency of alleles. For example, <a class="reference internal" href="refManual_ch3_sec2.html#InitGenotype" title="InitGenotype"><code class="xref py py-class docutils literal notranslate"><span class="pre">InitGenotype</span></code></a>(<code class="docutils literal notranslate"><span class="pre">freq=(0,</span> <span class="pre">0.2,</span>
<span class="pre">0.4,</span> <span class="pre">0.2)</span></code>) will assign allele 0, 1, 2, and 3 with probability 0, 0.2, 0.4 and
0.2 respectively.</li>
<li>By proportions of alleles. For example, <a class="reference internal" href="refManual_ch3_sec2.html#InitGenotype" title="InitGenotype"><code class="xref py py-class docutils literal notranslate"><span class="pre">InitGenotype</span></code></a>(<code class="docutils literal notranslate"><span class="pre">prop=(0,</span>
<span class="pre">0.2,</span> <span class="pre">0.4,</span> <span class="pre">0.2)</span></code>) will assign 400 allele 1, 800 allele 2 and 400 allele 3 to a
diploid population with 800 individuals.</li>
<li>By frequency of haplotypes. For example, <a class="reference internal" href="refManual_ch3_sec2.html#InitGenotype" title="InitGenotype"><code class="xref py py-class docutils literal notranslate"><span class="pre">InitGenotype</span></code></a>(<code class="docutils literal notranslate"><span class="pre">haplotypes=[[0,</span> <span class="pre">0],</span> <span class="pre">[1,1],</span> <span class="pre">[0,1],[1,1]]</span></code>) will assign four haplotypes with
equal probabilities. <a class="reference internal" href="refManual_ch3_sec2.html#InitGenotype" title="InitGenotype"><code class="xref py py-class docutils literal notranslate"><span class="pre">InitGenotype</span></code></a>(<code class="docutils literal notranslate"><span class="pre">haplotypes=[[0,</span> <span class="pre">0],</span> <span class="pre">[1,1],</span>
<span class="pre">[0,1],[1,1]],</span> <span class="pre">freq=[0.2,</span> <span class="pre">0.2,</span> <span class="pre">0.3,</span> <span class="pre">0.3]</span></code>) will assign these haplotypes with
different frequencies. If there are more than two loci, the haplotypes will be
repeated.</li>
<li>By frequency of haplotypes. For example, <a class="reference internal" href="refManual_ch3_sec2.html#InitGenotype" title="InitGenotype"><code class="xref py py-class docutils literal notranslate"><span class="pre">InitGenotype</span></code></a>(<code class="docutils literal notranslate"><span class="pre">haplotypes=[[0,</span> <span class="pre">0],</span> <span class="pre">[1,1],</span> <span class="pre">[0,1],[1,1]],</span> <span class="pre">prop=[0.2,</span> <span class="pre">0.2,</span> <span class="pre">0.3,</span> <span class="pre">0.3]</span></code>) will
assign four haplotypes with exact proportions.</li>
<li>By a list of genotype. For example, <a class="reference internal" href="refManual_ch3_sec2.html#InitGenotype" title="InitGenotype"><code class="xref py py-class docutils literal notranslate"><span class="pre">InitGenotype</span></code></a>(<code class="docutils literal notranslate"><span class="pre">genotype=[1,</span> <span class="pre">2,</span>
<span class="pre">2,</span> <span class="pre">1]</span></code>) will assign genotype <code class="docutils literal notranslate"><span class="pre">1</span></code>, <code class="docutils literal notranslate"><span class="pre">2</span></code>, <code class="docutils literal notranslate"><span class="pre">2</span></code>, <code class="docutils literal notranslate"><span class="pre">1</span></code> repeatedly to a
population. If individuals in this population has two homologous copies of a
chromosome with two loci, this operator will assign haplotype <code class="docutils literal notranslate"><span class="pre">1</span></code>, <code class="docutils literal notranslate"><span class="pre">2</span></code> to
the first homologous copy of the chromosome, and <code class="docutils literal notranslate"><span class="pre">2</span></code>, <code class="docutils literal notranslate"><span class="pre">1</span></code> to the second
copy.</li>
<li>By multiple allele frequencies or proportions returned by a function passed to
parameter <code class="docutils literal notranslate"><span class="pre">freq</span></code> or <code class="docutils literal notranslate"><span class="pre">prop</span></code> (new in version 1.1.7). This function can accept
parameters <code class="docutils literal notranslate"><span class="pre">loc</span></code>, <code class="docutils literal notranslate"><span class="pre">subPop</span></code> or <code class="docutils literal notranslate"><span class="pre">vsp</span></code> and returns locus, subpopopulation or
virtual subpopulation specific allele frequencies. For example, if you would
like to initialize genotypes with random allele frequency, you can set
<code class="docutils literal notranslate"><span class="pre">freq=lambda</span> <span class="pre">:</span> <span class="pre">random.random()</span></code> so that a new frequency is drawn from an
uniform distribution for each new locus. Note that simuPOP expects the return
value of this function to be a list of frequencies for alleles 0, 1, …, but
treats a single return value <em>x</em> as [<em>x, 1-x</em>] for simplicity.</li>
</ul>
<p>Parameter <code class="docutils literal notranslate"><span class="pre">loci</span></code> and <code class="docutils literal notranslate"><span class="pre">ploidy</span></code> can be used to specify a subset of loci and
homologous sets of chromosomes to initialize, and parameter <code class="docutils literal notranslate"><span class="pre">subPops</span></code> can be
used to specify subsets of individuals to initialize. Example <a class="reference internal" href="#initgenotype"><span class="std std-ref">InitGenotype</span></a> demonstrates how to use these the <a class="reference internal" href="refManual_ch3_sec2.html#InitGenotype" title="InitGenotype"><code class="xref py py-class docutils literal notranslate"><span class="pre">InitGenotype</span></code></a>
operator, including examples on how to define and use virtual subpopulations to
initialize individual genotype by sex or by proportion.</p>
<p id="initgenotype"><strong>Example</strong>: <em>Initialize individual genotype</em></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">simuPOP</span> <span class="k">as</span> <span class="nn">sim</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">sim</span><span class="o">.</span><span class="n">Population</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="p">[</span><span class="mi">2000</span><span class="p">,</span> <span class="mi">3000</span><span class="p">],</span> <span class="n">loci</span><span class="o">=</span><span class="p">[</span><span class="mi">5</span><span class="p">,</span> <span class="mi">7</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># by allele frequency</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">def</span> <span class="nf">printFreq</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="n">loci</span><span class="p">):</span>
<span class="gp">... </span>    <span class="n">sim</span><span class="o">.</span><span class="n">stat</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="n">alleleFreq</span><span class="o">=</span><span class="n">loci</span><span class="p">)</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s1">&#39;</span><span class="si">{:.3f}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">pop</span><span class="o">.</span><span class="n">dvars</span><span class="p">()</span><span class="o">.</span><span class="n">alleleFreq</span><span class="p">[</span><span class="n">x</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">loci</span><span class="p">]))</span>
<span class="gp">...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sim</span><span class="o">.</span><span class="n">initGenotype</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="n">freq</span><span class="o">=</span><span class="p">[</span><span class="o">.</span><span class="mi">4</span><span class="p">,</span> <span class="o">.</span><span class="mi">6</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sim</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="nb">max</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">structure</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">SubPopulation 0 (), 2000 Individuals:</span>
<span class="go">   0: MU 11000 0011111 | 11111 0101110</span>
<span class="go">   1: MU 00000 1111111 | 11101 1111001</span>
<span class="go">   2: MU 10111 0111100 | 01111 1011111</span>
<span class="go">   3: MU 11011 1101010 | 11010 1011111</span>
<span class="go">   4: MU 11011 0011010 | 10011 1001110</span>
<span class="go">   5: MU 00001 1010011 | 11111 1111110</span>
<span class="go">SubPopulation 1 (), 3000 Individuals:</span>
<span class="go">2000: MU 10011 0010100 | 01001 0011010</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">printFreq</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="nb">range</span><span class="p">(</span><span class="mi">5</span><span class="p">))</span>
<span class="go">0.397, 0.404, 0.400, 0.402, 0.406</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># by proportion</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sim</span><span class="o">.</span><span class="n">initGenotype</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="n">prop</span><span class="o">=</span><span class="p">[</span><span class="mf">0.4</span><span class="p">,</span> <span class="mf">0.6</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">printFreq</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="nb">range</span><span class="p">(</span><span class="mi">5</span><span class="p">))</span>
<span class="go">0.400, 0.400, 0.400, 0.400, 0.400</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># by haplotype frequency</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sim</span><span class="o">.</span><span class="n">initGenotype</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="n">freq</span><span class="o">=</span><span class="p">[</span><span class="o">.</span><span class="mi">4</span><span class="p">,</span> <span class="o">.</span><span class="mi">6</span><span class="p">],</span> <span class="n">haplotypes</span><span class="o">=</span><span class="p">[[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sim</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="nb">max</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">structure</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">SubPopulation 0 (), 2000 Individuals:</span>
<span class="go">   0: MU 11011 1011101 | 00100 1001001</span>
<span class="go">   1: MU 11011 1011101 | 11011 1011101</span>
<span class="go">   2: MU 00100 1001001 | 00100 1001001</span>
<span class="go">   3: MU 00100 1001001 | 00100 1001001</span>
<span class="go">   4: MU 11011 1011101 | 11011 1011101</span>
<span class="go">   5: MU 00100 1001001 | 11011 1011101</span>
<span class="go">SubPopulation 1 (), 3000 Individuals:</span>
<span class="go">2000: MU 00100 1001001 | 00100 1001001</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">printFreq</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="nb">range</span><span class="p">(</span><span class="mi">5</span><span class="p">))</span>
<span class="go">0.597, 0.597, 0.403, 0.597, 0.597</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># by haplotype proportion</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sim</span><span class="o">.</span><span class="n">initGenotype</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="n">prop</span><span class="o">=</span><span class="p">[</span><span class="mf">0.4</span><span class="p">,</span> <span class="mf">0.6</span><span class="p">],</span> <span class="n">haplotypes</span><span class="o">=</span><span class="p">[[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">printFreq</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="nb">range</span><span class="p">(</span><span class="mi">5</span><span class="p">))</span>
<span class="go">0.600, 0.600, 0.400, 0.000, 0.600</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># by genotype</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">sim</span><span class="o">.</span><span class="n">Population</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">],</span> <span class="n">loci</span><span class="o">=</span><span class="p">[</span><span class="mi">5</span><span class="p">,</span> <span class="mi">7</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sim</span><span class="o">.</span><span class="n">initGenotype</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="n">genotype</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">*</span><span class="mi">5</span> <span class="o">+</span> <span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">*</span><span class="mi">7</span> <span class="o">+</span> <span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">*</span><span class="mi">5</span> <span class="o">+</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span><span class="o">*</span><span class="mi">7</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sim</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="n">structure</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">SubPopulation 0 (), 2 Individuals:</span>
<span class="go">   0: MU 11111 2222222 | 33333 4444444</span>
<span class="go">   1: MU 11111 2222222 | 33333 4444444</span>
<span class="go">SubPopulation 1 (), 3 Individuals:</span>
<span class="go">   2: MU 11111 2222222 | 33333 4444444</span>
<span class="go">   3: MU 11111 2222222 | 33333 4444444</span>
<span class="go">   4: MU 11111 2222222 | 33333 4444444</span>

<span class="gp">&gt;&gt;&gt; </span><span class="c1">#</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># use virtual subpopulation</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">sim</span><span class="o">.</span><span class="n">Population</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="p">[</span><span class="mi">2000</span><span class="p">,</span> <span class="mi">3000</span><span class="p">],</span> <span class="n">loci</span><span class="o">=</span><span class="p">[</span><span class="mi">5</span><span class="p">,</span> <span class="mi">7</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span><span class="o">.</span><span class="n">setVirtualSplitter</span><span class="p">(</span><span class="n">sim</span><span class="o">.</span><span class="n">SexSplitter</span><span class="p">())</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sim</span><span class="o">.</span><span class="n">initSex</span><span class="p">(</span><span class="n">pop</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sim</span><span class="o">.</span><span class="n">initGenotype</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="n">genotype</span><span class="o">=</span><span class="nb">range</span><span class="p">(</span><span class="mi">10</span><span class="p">),</span> <span class="n">loci</span><span class="o">=</span><span class="nb">range</span><span class="p">(</span><span class="mi">5</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># initialize all males</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sim</span><span class="o">.</span><span class="n">initGenotype</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="n">genotype</span><span class="o">=</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">*</span><span class="mi">7</span><span class="p">,</span> <span class="n">loci</span><span class="o">=</span><span class="nb">range</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">12</span><span class="p">),</span>
<span class="gp">... </span>    <span class="n">subPops</span><span class="o">=</span><span class="p">[(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># assign genotype by proportions</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span><span class="o">.</span><span class="n">setVirtualSplitter</span><span class="p">(</span><span class="n">sim</span><span class="o">.</span><span class="n">ProportionSplitter</span><span class="p">([</span><span class="mf">0.4</span><span class="p">,</span> <span class="mf">0.6</span><span class="p">]))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sim</span><span class="o">.</span><span class="n">initGenotype</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="n">freq</span><span class="o">=</span><span class="p">[</span><span class="mf">0.2</span><span class="p">,</span> <span class="mf">0.8</span><span class="p">],</span> <span class="n">subPops</span><span class="o">=</span><span class="p">[(</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sim</span><span class="o">.</span><span class="n">initGenotype</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="n">freq</span><span class="o">=</span><span class="p">[</span><span class="mf">0.5</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">],</span> <span class="n">subPops</span><span class="o">=</span><span class="p">[(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># initialize by random allele frequency</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">random</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sim</span><span class="o">.</span><span class="n">initGenotype</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="n">freq</span><span class="o">=</span><span class="k">lambda</span> <span class="p">:</span> <span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">())</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">printFreq</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="nb">range</span><span class="p">(</span><span class="mi">5</span><span class="p">))</span>
<span class="go">0.580, 0.239, 0.100, 0.576, 0.674</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># initialize with loci specific frequency. here</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># lambda loc: 0.01*loc is equivalent to</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># lambda loc: [0.01*loc, 1-0.01*loc]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sim</span><span class="o">.</span><span class="n">initGenotype</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">freq</span><span class="o">=</span><span class="k">lambda</span> <span class="n">loc</span><span class="p">:</span> <span class="mf">0.01</span><span class="o">*</span><span class="n">loc</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">printFreq</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="nb">range</span><span class="p">(</span><span class="mi">5</span><span class="p">))</span>
<span class="go">0.000, 0.009, 0.018, 0.029, 0.041</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># initialize with VSP-specific frequency</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sim</span><span class="o">.</span><span class="n">initGenotype</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">freq</span><span class="o">=</span><span class="k">lambda</span> <span class="n">vsp</span><span class="p">:</span> <span class="p">[[</span><span class="mf">0.2</span><span class="p">,</span> <span class="mf">0.8</span><span class="p">],</span> <span class="p">[</span><span class="mf">0.5</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">]][</span><span class="n">vsp</span><span class="p">[</span><span class="mi">1</span><span class="p">]],</span>
<span class="gp">... </span>    <span class="n">subPops</span><span class="o">=</span><span class="p">[(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)])</span>
<span class="go">&gt;&gt;&gt;</span>

<span class="go">now exiting runScriptInteractively...</span>
</pre></div>
</div>
<p><a class="reference external" href="InitGenotype.py">Download InitGenotype.py</a></p>
</div>
<div class="section" id="initialize-information-fields-operator-initinfo">
<h2>Initialize information fields (operator <a class="reference internal" href="refManual_ch3_sec2.html#InitInfo" title="InitInfo"><code class="xref py py-class docutils literal notranslate"><span class="pre">InitInfo</span></code></a>)<a class="headerlink" href="#initialize-information-fields-operator-initinfo" title="Permalink to this headline">¶</a></h2>
<p>Operator <a class="reference internal" href="refManual_ch3_sec2.html#InitInfo" title="InitInfo"><code class="xref py py-class docutils literal notranslate"><span class="pre">InitInfo</span></code></a> and its function form <code class="docutils literal notranslate"><span class="pre">initInfo</span></code> initialize one or
more information fields of all individuals or Individuals in selected (virtual)
subpopulations using either a list of values or a Python function. If a value or
a list of value is given, it will be used repeatedly to assign values of
specified information fields of all applicable individuals. For example,
<code class="docutils literal notranslate"><span class="pre">initInfo(pop,</span> <span class="pre">values=1,</span> <span class="pre">infoFields='x')</span></code> will assign value <code class="docutils literal notranslate"><span class="pre">1</span></code> to
information field <code class="docutils literal notranslate"><span class="pre">x</span></code> of all individuals, and</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">initInfo</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">],</span> <span class="n">infoFields</span><span class="o">=</span><span class="s1">&#39;x&#39;</span><span class="p">,</span> <span class="n">subPops</span><span class="o">=</span><span class="p">[(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">)])</span>
</pre></div>
</div>
<p>will assign values <code class="docutils literal notranslate"><span class="pre">1</span></code>, <code class="docutils literal notranslate"><span class="pre">2</span></code>, <code class="docutils literal notranslate"><span class="pre">3</span></code>, <code class="docutils literal notranslate"><span class="pre">1</span></code>, <code class="docutils literal notranslate"><span class="pre">2</span></code>, <code class="docutils literal notranslate"><span class="pre">3</span></code>… to information
field <code class="docutils literal notranslate"><span class="pre">x</span></code> of individuals in the second virtual subpopulation of subpopulation
0.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">values</span></code> parameter also accepts a Python function. This feature is usually
used to assign random values to an information field. For example,
<code class="docutils literal notranslate"><span class="pre">values=random.random</span></code> would assign a random value between 0 and 1. If a
function takes parameters, a lambda function can be used. For example,</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">initInfo</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="k">lambda</span> <span class="p">:</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">5</span><span class="p">),</span> <span class="n">infoFields</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;x&#39;</span><span class="p">,</span> <span class="s1">&#39;y&#39;</span><span class="p">])</span>
</pre></div>
</div>
<p>assigns random integers between 2 and 5 to information fields <code class="docutils literal notranslate"><span class="pre">x</span></code> and <code class="docutils literal notranslate"><span class="pre">y</span></code> of
all individuals in <em>pop</em>. Example <a class="reference internal" href="#initinfo"><span class="std std-ref">InitInfo</span></a> demonstrates these
usages.</p>
<p id="initinfo"><strong>Example</strong>: <em>initialize information fields</em></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">simuPOP</span> <span class="k">as</span> <span class="nn">sim</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">random</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">sim</span><span class="o">.</span><span class="n">Population</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="p">[</span><span class="mi">5</span><span class="p">],</span> <span class="n">loci</span><span class="o">=</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="n">infoFields</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;sex&#39;</span><span class="p">,</span> <span class="s1">&#39;age&#39;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span><span class="o">.</span><span class="n">setVirtualSplitter</span><span class="p">(</span><span class="n">sim</span><span class="o">.</span><span class="n">SexSplitter</span><span class="p">())</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sim</span><span class="o">.</span><span class="n">initSex</span><span class="p">(</span><span class="n">pop</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sim</span><span class="o">.</span><span class="n">initInfo</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="n">subPops</span><span class="o">=</span><span class="p">[(</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">)],</span> <span class="n">infoFields</span><span class="o">=</span><span class="s1">&#39;sex&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sim</span><span class="o">.</span><span class="n">initInfo</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">subPops</span><span class="o">=</span><span class="p">[(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">)],</span> <span class="n">infoFields</span><span class="o">=</span><span class="s1">&#39;sex&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sim</span><span class="o">.</span><span class="n">initInfo</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="k">lambda</span><span class="p">:</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span> <span class="mi">70</span><span class="p">),</span> <span class="n">infoFields</span><span class="o">=</span><span class="s1">&#39;age&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sim</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="n">structure</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">SubPopulation 0 (), 5 Individuals:</span>
<span class="go">   0: FU 00 | 00 |  1 39</span>
<span class="go">   1: FU 00 | 00 |  1 29</span>
<span class="go">   2: MU 00 | 00 |  0 68</span>
<span class="go">   3: MU 00 | 00 |  0 50</span>
<span class="go">   4: MU 00 | 00 |  0 21</span>


<span class="go">now exiting runScriptInteractively...</span>
</pre></div>
</div>
<p><a class="reference external" href="InitInfo.py">Download InitInfo.py</a></p>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">simuPOP</a></h1>








<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="userGuide_ch1.html">Front Matter</a></li>
<li class="toctree-l1"><a class="reference internal" href="userGuide_ch2.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="userGuide_ch3.html">Loading and running simuPOP</a></li>
<li class="toctree-l1"><a class="reference internal" href="userGuide_ch4.html">Individuals and Populations</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="userGuide_ch5.html">simuPOP Operators</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch5_sec1.html">Introduction to operators</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Initialization</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#initialize-individual-sex-operator-initsex">Initialize individual sex (operator <code class="docutils literal notranslate"><span class="pre">InitSex</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#initialize-genotype-operator-initgenotype">Initialize genotype (operator <code class="docutils literal notranslate"><span class="pre">InitGenotype</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#initialize-information-fields-operator-initinfo">Initialize information fields (operator <code class="docutils literal notranslate"><span class="pre">InitInfo</span></code>)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch5_sec3.html">Expressions and statements</a></li>
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch5_sec4.html">Demographic changes</a></li>
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch5_sec5.html">Genotype transmitters</a></li>
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch5_sec6.html">Mutation</a></li>
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch5_sec7.html">Penetrance</a></li>
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch5_sec8.html">Quantitative trait</a></li>
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch5_sec9.html">Natural Selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch5_sec10.html">Tagging operators</a></li>
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch5_sec11.html">Statistics calculation (operator <code class="docutils literal notranslate"><span class="pre">Stat</span></code>)</a></li>
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch5_sec12.html">Conditional operators</a></li>
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch5_sec13.html">Miscellaneous operators</a></li>
<li class="toctree-l2"><a class="reference internal" href="userGuide_ch5_sec14.html">Hybrid and Python operators</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="userGuide_ch6.html">Evolving populations</a></li>
<li class="toctree-l1"><a class="reference internal" href="userGuide_ch7.html">Utility Modules</a></li>
<li class="toctree-l1"><a class="reference internal" href="userGuide_ch8.html">A real world example</a></li>
<li class="toctree-l1"><a class="reference internal" href="refManual_ch1.html">Front Matter</a></li>
<li class="toctree-l1"><a class="reference internal" href="refManual_ch2.html">simuPOP Components</a></li>
<li class="toctree-l1"><a class="reference internal" href="refManual_ch3.html">Operator References</a></li>
<li class="toctree-l1"><a class="reference internal" href="refManual_ch4.html">Utility Modules</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="userGuide.html">&lt;no title&gt;</a><ul>
  <li><a href="userGuide_ch5.html">simuPOP Operators</a><ul>
      <li>Previous: <a href="userGuide_ch5_sec1.html" title="previous chapter">Introduction to operators</a></li>
      <li>Next: <a href="userGuide_ch5_sec3.html" title="next chapter">Expressions and statements</a></li>
  </ul></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2004-2019, Bo Peng.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.8.4</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/userGuide_ch5_sec2.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>