
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=cp1252" />
    <title>Score Cells</title>
    <link rel="stylesheet" href="../_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/style.css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="next" title="Filter" href="filter.html" />
    <link rel="prev" title="Score Genes" href="scoregenes.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="filter.html" title="Filter"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="scoregenes.html" title="Score Genes"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">scOrange 0.5 documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
          <div class="body" role="main">
            
  <div class="section" id="score-cells">
<h1>Score Cells</h1>
<p>Add a cell score based on the given set of genes.</p>
<dl class="docutils">
<dt>Inputs</dt>
<dd><dl class="first last docutils">
<dt>Data</dt>
<dd>input dataset</dd>
<dt>Genes</dt>
<dd>dataset with gene names</dd>
</dl>
</dd>
<dt>Outputs</dt>
<dd><dl class="first last docutils">
<dt>Data</dt>
<dd>dataset with a meta column reporting on cell scores</dd>
</dl>
</dd>
</dl>
<p><strong>Score Cells</strong> scores the cells (rows) in the input data based on expression of input marker genes. Input data has to include gene names in column headers (as attribute names). The score is computed independently for each cell and is equal to the maximum expression of the marker genes. If expressions for all the marker genes are missing, the cell obtains a score of zero. If <em>Genes</em> input is provided, one of the columns must include names of genes that match those from expression data.</p>
<div class="figure">
<img alt="../_images/ScoreCells-stamped.png" src="../_images/ScoreCells-stamped.png" />
</div>
<ol class="arabic simple">
<li>A column name from table on genes that provides names of genes that are included in the input gene expression table.</li>
<li>Tick to automatically process input data and send the result of scoring to the output. If left unchecked, processing must be triggered manually.</li>
</ol>
<div class="section" id="example">
<h2>Example</h2>
<p>We will use <a class="reference internal" href="singlecelldatasets.html"><span class="doc">Single Cell Datasets</span></a> widget to load <em>Bone marrow mononuclear cells with AML (sample)</em> data. Then we will pass it through <strong>k-Means</strong> and select 2 clusters from Silhouette Scores. Ok, it looks like there might be two distinct clusters here.</p>
<p>But can we find subpopulations in these cells? Let us load <em>Bone marrow mononuclear cells with AML (markers)</em> with <a class="reference internal" href="singlecelldatasets.html"><span class="doc">Single Cell Datasets</span></a>. Now, pass the marker genes to <strong>Data Table</strong> and select, for example, natural killer cells from the list (NKG7).</p>
<p>Pass the markers and k-Means results to <strong>Score Cells</strong> widget and select <em>geneName</em> to match markers with genes. Finally, add <a class="reference internal" href="tsne.html"><span class="doc">t-SNE</span></a> to visualize the results.</p>
<p>In <strong>t-SNE</strong>, use <em>Scores</em> attribute to color the points and set their size. We see that killer cells are nicely clustered together and that t-SNE indeed found subpopulations.</p>
<div class="figure">
<img alt="../_images/tSNE-Example.png" src="../_images/tSNE-Example.png" />
</div>
</div>
</div>


          </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="filter.html" title="Filter"
             >next</a> |</li>
        <li class="right" >
          <a href="scoregenes.html" title="Score Genes"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">scOrange 0.5 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2018, Biolab.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.8.1.
    </div>
  </body>
</html>